---
title: "University Lake Microbial Seed Bank"
author: "Nathan Wisnoski, Jay Lennon"
date: "1/19/2018"
output: html_document
---

```{r setup, include=FALSE}
# Load packages

require(vegan)
require(tidyverse)
require(lubridate)
require(codyn)
require(viridis)
require(betapart)
require(cowplot)
#require(xts)
require(progress)
#require(cluster)
#require(adespatial)
require(ggrepel)

source("bin/mothur_tools.R")

cv <- function(x){
  sd(x) / abs(mean(x))
}

set.seed(47405)
```

Next, we'll set the aesthetics of the figures we will produce. 
```{r figure_setup}
#my.cols <- RColorBrewer::brewer.pal(n = 4, name = "Greys")[3:4]

# Set theme for figures in the paper
theme_set(theme_classic() + 
  theme(axis.title = element_text(size = 16),
        axis.title.x = element_text(margin = margin(t = 15, b = 15)),
        axis.title.y = element_text(margin = margin(l = 15, r = 15)),
        axis.text = element_text(size = 14),
        axis.text.x = element_text(margin = margin(t = 5)),
        axis.text.y = element_text(margin = margin(r = 5)),
        #axis.line.x = element_line(size = 1),
        #axis.line.y = element_line(size = 1),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_line(size = 1),
        axis.ticks.y = element_line(size = 1),
        axis.ticks.length = unit(.1, "in"),
        panel.border = element_rect(color = "black", fill = NA, size = 1.5),
        legend.title = element_blank(),
        legend.text = element_text(size = 14),
        strip.text = element_text(size = 14),
        strip.background = element_blank()
        ))
```

## Import Data
Here, we read in the processed sequence files from mothur (shared and taxonomy) and a design of the sampling. We also load in the environmental data. We then remove the mock community from the dataset and ensure the the design and OTU table are aligned by row.
```{r}
# Define Inputs
# Design = general design file for experiment
# shared = OTU table from mothur with sequence similarity clustering
# Taxonomy = Taxonomic information for each OTU
design <- "data/UL_timeseries_design.csv"
shared <- "data/ul_resgrad.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.pick.opti_mcc.shared"
taxon  <- "data/ul_resgrad.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.pick.opti_mcc.0.03.cons.taxonomy"

# Import Design
design <- read_csv(file = design)

# Import Shared Files
OTUs <- read.otu(shared = shared, cutoff = "0.03")    # 97% Similarity

# Import Taxonomy
OTU.tax <- read.tax(taxonomy = taxon, format = "rdp")


# Load environmental data
env.data <- read.table("data/ul-seedbank.env.txt", sep="\t", header=TRUE)
env.data$date <- as.Date(parse_date_time(env.data$date, "m d y"))
env.data$doc[which(env.data$doc == "**")] <- NA
env.data$doc <- as.numeric(env.data$doc)
summary(env.data)
design <- left_join(design, env.data[,c("sample.id", "date")])
env.data <- env.data[-which(!(env.data$date %in% design$date)),]

# Subset to just the reservoir gradient sites
OTUs <- OTUs[str_which(rownames(OTUs), "UL"),]

# make sure OTU table matches up with design order
OTUs <- OTUs[match(design$sample.name, rownames(OTUs)),]
```

# transform OTUs
```{r data format, include=FALSE}

# Make rel abund matrices and split into active total comms
# OTUs <- OTUs[,-which(colSums(OTUs) < 5)]
OTUs <- rrarefy(OTUs, sample = min(rowSums(OTUs)))

OTUs.active <- OTUs[which(design$sample.type == "RNA"),]
OTUs.total <- OTUs[which(design$sample.type == "DNA"),]
OTUs.total <- decostand(OTUs.active, method = "pa") + OTUs.total
OTUs.total.pa <- decostand(OTUs.total, method = "pa")
OTUs.active.pa <- decostand(OTUs.active, method = "pa")

OTUs.REL <- decostand(OTUs, method = "hellinger")
#OTUs.REL <- decostand(OTUs, method = "total")
OTUs.REL.total <- OTUs.REL[which(design$sample.type == "DNA"),]
OTUs.REL.active <- OTUs.REL[which(design$sample.type == "RNA"),]


# Define color pallette
my.cols <- viridis(4, begin = .3, end = 0.8)
```

# Characterize temporal dynamics of the bacterioplankton community 

## How does alpha diversity vary over time?
```{r}
alpha.div.total <- exp(diversity(OTUs.total, index = "shannon")) # spec equivs
alpha.div.active <- exp(diversity(OTUs.active, index = "shannon"))
data.frame(date = env.data$date, Total = alpha.div.total, Active = alpha.div.active) %>% 
  gather(community, diversity, -date) %>% 
  ggplot(aes(x = date, y = diversity, color = community)) + 
  geom_vline(xintercept = as.numeric(as.Date(c("2014-01-01", "2015-01-01"))),
             linetype = 2, color = "grey", alpha = 0.8) +
  geom_line(size = 1, alpha = .9, show.legend = F) +
  ylab(expression(paste(alpha,"-diversity (species equivalents)"))) +
  xlab("Date") +
  scale_x_date(date_breaks = "3 month", 
               date_minor_breaks = "1 month", 
               date_labels = "%b %Y") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  #scale_color_manual("Community Subset", values = c("#b2df8a", "#1f78b4")) +
  scale_color_manual(values = my.cols[c(4,1)]) +
  scale_y_log10(limits = c(1,1000),
                breaks = c(1,10,100,1000),
                labels = c("1", "10", "100", "1000")) +
  annotation_logticks(sides = "l") +
  ggsave("figures/alpha_diversity.pdf", bg = "white", height = 6, width = 8)
```


## How does community composition change over time?
### Ordination methods (abundance-based)
```{r}
# never active
transients <- OTUs.REL.total[,which(colSums(OTUs.REL.active) == 0 & colSums(OTUs.REL.total) != 0)]
transients.dist <- vegdist(transients, method = "euclidean")
transients.pcoa <- cmdscale(transients.dist, eig = T)
var1 <- round(eigenvals(transients.pcoa)[1] / sum(eigenvals(transients.pcoa)), 3) * 100
var2 <- round(eigenvals(transients.pcoa)[2] / sum(eigenvals(transients.pcoa)), 3) * 100
transients.traj <- transients.pcoa$points
colnames(transients.traj) <- c("Comp.1", "Comp.2")
transients.traj.df <- cbind.data.frame(
  sample.id = design$sample.id[which(design$sample.type == "RNA")], transients.traj) 


filter(design, sample.type == "RNA") %>% 
  full_join(transients.traj.df) %>% 
  mutate(month = lubridate::month(date, label = T, abbr = T)) %>% 
  ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
  geom_point() + 
  scale_color_viridis(discrete = T, "Month") + 
  theme_cowplot() + 
  xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) + 
  coord_fixed() +
  stat_ellipse()


# present, but not active
seedbank <- OTUs.REL.active == 0 & OTUs.REL.total != 0
seedbank <- seedbank * OTUs.REL.total
seedbank <- seedbank[,which(colSums(seedbank) > 0)]
# remove transients
seedbank <- seedbank[, -which(colnames(seedbank) %in% colnames(transients))]

seedbank.dist <- vegdist(seedbank, method = "euclidean")
seedbank.pcoa <- cmdscale(seedbank.dist, eig = T)
var1 <- round(eigenvals(seedbank.pcoa)[1] / sum(eigenvals(seedbank.pcoa)), 3) * 100
var2 <- round(eigenvals(seedbank.pcoa)[2] / sum(eigenvals(seedbank.pcoa)), 3) * 100
seedbank.traj <- seedbank.pcoa$points
colnames(seedbank.traj) <- c("Comp.1", "Comp.2")

seedbank.traj.df <- cbind.data.frame(
  sample.id = design$sample.id[which(design$sample.type == "RNA")], seedbank.traj) 
filter(design, sample.type == "RNA") %>% 
  full_join(seedbank.traj.df) %>% 
  mutate(month = lubridate::month(date, label = T, abbr = T)) %>%
  ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
  geom_point(alpha = 0.5) +
  # geom_path(arrow = arrow(angle = 15, length = unit(0.1, "inches"),
  #                         ends = "last", type = "closed")) + 
  scale_color_viridis(discrete = T, "Month") + 
  geom_point(aes(x = seedbank.traj.df$Comp.1[1], 
                 y = seedbank.traj.df$Comp.2[1]),
             color = viridis(2)[1], size = 3) +
  geom_point(aes(x = seedbank.traj.df$Comp.1[123], 
                 y = seedbank.traj.df$Comp.2[123]),
             color = viridis(2)[2], size = 3) +
  theme_cowplot() + 
  xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) +
  coord_fixed() +
  stat_ellipse() +
  ggsave("figures/seedbank_trajectory.pdf", bg = "white", height = 8, width = 10)

filter(design, sample.type == "RNA") %>% 
  full_join(seedbank.traj.df) %>% 
  mutate(month = lubridate::month(date, label = T, abbr = T)) %>% 
  ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
  geom_point() + 
  scale_color_viridis(discrete = T, "Month") + 
  theme_cowplot() + 
  xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) + 
  stat_ellipse() +
  coord_fixed() +
  ggsave("figures/seedbank_pcoa.pdf", bg = "white", height = 8, width = 10)


# seedbank.groups <- filter(design, sample.type == "RNA") %>% 
#   full_join(seedbank.traj.df) %>% 
#   mutate(month = month(date, label = T, abbr = T)) %>% 
#   select(month)
# seedbank.groups <- factor(seedbank.groups$month)
# sb.bdisp <- betadisper(seedbank.dist, group = seedbank.groups)
# plot(sb.bdisp)
# anova(sb.bdisp)
# boxplot(sb.bdisp)

OTUs.dist <- vegdist(OTUs.REL, method = "euclidean")
OTUs.pcoa <- cmdscale(OTUs.dist, eig = T)
var1 <- round(eigenvals(OTUs.pcoa)[1] / sum(eigenvals(OTUs.pcoa)), 3) * 100
var2 <- round(eigenvals(OTUs.pcoa)[2] / sum(eigenvals(OTUs.pcoa)), 3) * 100
OTUs.traj <- OTUs.pcoa$points
colnames(OTUs.traj) <- c("Comp.1", "Comp.2")
active.traj <- OTUs.traj[which(design$sample.type == "RNA"),]
total.traj <- OTUs.traj[which(design$sample.type == "DNA"),]

active.traj.df <- cbind.data.frame(
  sample.id = design$sample.id[which(design$sample.type == "RNA")], active.traj) 

full_join(design[which(design$sample.type == "RNA"),], active.traj.df) %>%
  ggplot(aes(x = Comp.1, y = Comp.2, color = sample.id)) + 
  geom_path(arrow = arrow(angle = 15, length = unit(0.1, "inches"),
                  ends = "last", type = "closed")) + 
  scale_color_gradientn(colors = viridis(125), "Sample Number") + 
  geom_point(aes(x = active.traj.df$Comp.1[1], y = active.traj.df$Comp.2[1]),
             color = viridis(2)[1], size = 3) +
  geom_point(aes(x = active.traj.df$Comp.1[123], y = active.traj.df$Comp.2[123]),
             color = viridis(2)[2], size = 3) +
  theme_cowplot() + 
  xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) +
  coord_fixed() +
  ggsave("figures/active_trajectory.pdf", bg = "white", height = 8, width = 10)

total.traj.df <- cbind.data.frame(
  sample.id = design$sample.id[which(design$sample.type == "DNA")], total.traj) 
full_join(design[which(design$sample.type == "DNA"),], total.traj.df) %>%
  ggplot(aes(x = Comp.1, y = Comp.2, color = sample.id)) + 
  geom_path(arrow = arrow(angle = 15, length = unit(0.1, "inches"),
                          ends = "last", type = "closed")) + 
  scale_color_gradientn(colors = magma(125)) + 
  geom_point(aes(x = total.traj.df$Comp.1[1], y = total.traj.df$Comp.2[1]),
             color = magma(2)[1], size = 3) +
  geom_point(aes(x = total.traj.df$Comp.1[123], y = total.traj.df$Comp.2[123]),
             color = magma(2)[2], size = 3) +
  theme_cowplot()+
  xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) + 
  coord_fixed() +
  ggsave("figures/total_trajectory.pdf", bg = "white", height = 8, width = 10)


# Points colored by month
# Active
filter(design, sample.type == "RNA") %>% 
  full_join(active.traj.df) %>% 
  mutate(month = lubridate::month(date, label = T, abbr = T)) %>% 
  ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
  geom_point() + 
  scale_color_viridis(discrete = T, "Month") + 
  theme_cowplot() + 
  xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) +
  stat_ellipse() +
  coord_fixed() +
  ggsave("figures/active_pcoa.pdf", bg = "white", height = 8, width = 10)

# total
filter(design, sample.type == "DNA") %>% 
  full_join(total.traj.df) %>% 
  mutate(month = lubridate::month(date, label = T, abbr = T)) %>% 
  ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
  geom_point() + 
  scale_color_viridis(discrete = T, "Month") + 
  theme_cowplot() + 
  xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) +
  stat_ellipse() +
  coord_fixed() +
  ggsave("figures/total_pcoa.pdf", bg = "white", height = 8, width = 10)

pc1.dyn <- rbind.data.frame(cbind.data.frame(total.traj.df, subset = "Total", Date = env.data$date), 
                 cbind.data.frame(active.traj.df, subset = "Active", Date = env.data$date)) %>% 
  ggplot(aes(x = Date, y = Comp.1, color = subset)) + 
  geom_line(size = 0.75) + 
  scale_color_manual("Community Subset", values = my.cols[c(1,4)]) + 
  ylab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
  xlab("") +
  scale_x_date(date_breaks = "2 month", date_minor_breaks = "1 month", date_labels = "%b-%y") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  theme(legend.direction = "vertical", legend.position = "right")

pc2.dyn <- rbind.data.frame(cbind.data.frame(total.traj.df, subset = "Total", Date = env.data$date), 
                 cbind.data.frame(active.traj.df, subset = "Active", Date = env.data$date)) %>% 
  ggplot(aes(x = Date, y = Comp.2, color = subset)) + 
  geom_line(size = 0.75) + 
  scale_color_manual("Community Subset", values = my.cols[c(1,4)]) + 
  ylab(paste("PCoA 2 (",var2,"%)", sep = "")) + 
  xlab("Date") +
  scale_x_date(date_breaks = "2 month", date_minor_breaks = "1 month", date_labels = "%b-%y") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1))

plots <- plot_grid(pc1.dyn + theme(legend.position = "none"), 
          pc2.dyn + theme(legend.position = "none"), nrow = 2, align = 'hv', axis = 'l',
          labels = c("A", "B"))
plots + ggsave(filename = "figures/pcoa_dynamics.pdf", width = 8, height = 8, units = "in")
apply(total.traj, 2, sd)
apply(active.traj, 2, sd)
```


### Turnover Analysis
```{r}
OTUs.long <- as.data.frame(OTUs)
OTUs.long$mol <- design$sample.type
OTUs.long$number <- design$sample.id
OTUs.long <- gather(OTUs.long, key = otu, value = abundance, -mol, -number)


# Calculate turnover
turn.total <- turnover(df = OTUs.long, 
                 time.var = "number",
                 species.var = "otu",
                 abundance.var = "abundance",
                 replicate.var = "mol")
turn.appear <- turnover(df = OTUs.long, 
                       time.var = "number",
                       species.var = "otu",
                       abundance.var = "abundance",
                       replicate.var = "mol",
                       metric = "appearance")
turn.disappear <- turnover(df = OTUs.long, 
                       time.var = "number",
                       species.var = "otu",
                       abundance.var = "abundance",
                       replicate.var = "mol",
                       metric = "disappearance")

# Make figure showing turnover in different subsets
turn.total %>% mutate(Molecule = ifelse(turn.total$mol == "DNA", "Total", "Active")) %>%
  ggplot(aes(y = total, x = number, color = Molecule))+
  geom_point(size = .75, alpha = 0.5) + 
  geom_smooth(method = "loess") + 
  ylab("Turnover") + 
  xlab("Time (Week)") +
  scale_color_manual("Community Subset", values = my.cols[c(4,1)]) +
  ggsave("figures/turnover-total.pdf", height = 6, width = 8, bg = "white")
turn.appear %>% mutate(Molecule = ifelse(turn.appear$mol == "DNA", "Total", "Active")) %>%
  ggplot(aes(y = appearance, x = number, color = Molecule))+
  geom_point(size = .75, alpha = 0.5) + 
  geom_smooth(method = "loess") + 
  ylab("Turnover") + 
  xlab("Time (Week)") +
  scale_color_manual("Community Subset", values = my.cols[c(4,1)]) +
  ggsave("figures/turnover-appear.pdf", height = 6, width = 8, bg = "white")
turn.disappear %>% mutate(Molecule = ifelse(turn.disappear$mol == "DNA", "Total", "Active")) %>%
  ggplot(aes(y = disappearance, x = number, color = Molecule))+
  geom_point(size = .75, alpha = 0.5) + 
  geom_smooth(method = "loess") + 
  ylab("Turnover") + 
  xlab("Time (Week)") +
  scale_color_manual("Community Subset", values = my.cols[c(4,1)]) +
  ggsave("figures/turnover-disappear.pdf", height = 6, width = 8, bg = "white")


dates <- left_join(design[-which(design$sample.id == 1),], env.data) %>% select(date)
turn.total$Season <- ifelse(month(dates$date) %in% c(3:5), "Spring", 
                            ifelse(month(dates$date) %in% c(6:8), "Summer",
                                   ifelse(month(dates$date) %in% c(9:11), "Fall",
                                          ifelse(month(dates$date) %in% c(12,1,2), "Winter", NA))))
group_by(turn.total, mol, Season) %>% 
  summarise(
    mean = mean(total),
    cv = sd(total)/mean)
turn.total %>% mutate(Molecule = ifelse(turn.total$mol == "DNA", "Total", "Active")) %>%
  ggplot(aes(y = total, fill = Molecule, x = Season))+
  geom_boxplot(alpha = 0.5) +
  scale_fill_manual("Community Subset", values = my.cols[c(4,1)]) +
  ylab("Turnover") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ggsave("figures/turnover_season.pdf", bg = "white", width = 6, height = 6)

turn.total %>% mutate(Molecule = ifelse(turn.total$mol == "DNA", "Total", "Active")) %>%
  mutate(Month = month(dates$date, label = T, abbr = T)) %>% 
  ggplot(aes(y = total, fill = Molecule, x = Month)) +
  geom_boxplot(alpha = 0.5) +
  scale_fill_manual("Community Subset", values = my.cols[c(4,1)]) +
  ylab("Turnover") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) + 
  ggsave("figures/turnover_month.pdf", bg = "white", width = 8, height = 6)
```

### Mean Rank Shift
```{r}
# Mean Rank Shift

head(OTUs.long)
ul.mrs <- rank_shift(df = OTUs.long,
           time.var = "number",
           species.var = "otu",
           abundance.var = "abundance",
           replicate.var = "mol")

ul.mrs$year <- as.numeric(colsplit(ul.mrs$year_pair, pattern = "-", names = c("year1", "year2"))[,2])
names(ul.mrs)[3] <- "Molecule"


# Create plot
ul.mrs %>% mutate(Molecule = ifelse(Molecule == "DNA", "Total", "Active"), 
                  Date = env.data$date[year]) %>% 
  ggplot(aes(x = Date, y = MRS, color = Molecule)) + 
  geom_line(size = .6, alpha = 0.75) +
  xlab("Date") +
  scale_x_date(date_breaks = "2 month", date_minor_breaks = "1 month", date_labels = "%b-%y") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  ylab("Mean Rank Shift") +
  # scale_y_log10(limits = c(100,1000),
  #               breaks = c(100,1000),
  #               labels = c("100", "1000")) +
  annotation_logticks(sides = "l") +
  scale_color_manual("Community Subset", values = my.cols[c(4,1)]) +
  ggsave("figures/mean_rank_shift.pdf", bg = "white", height = 6, width = 8)



# Does one plot type show higher or lower MRS, on average?
dates <- left_join(design[-which(design$sample.id == 1),], env.data) %>% select(date)

ul.mrs$Season <- ifelse(month(dates$date) %in% c(5:10), "Summer", "Winter")
group_by(ul.mrs, Molecule, Season) %>% 
  summarise(
    mean = mean(MRS),
    cv = sd(MRS)/mean)


# # Rate change interval
# ul.rci <- rate_change_interval(df = OTUs.long,
#                      time.var = "number",
#                      species.var = "otu", 
#                      abundance.var = "abundance",
#                      replicate.var = "mol")
# 
# ul.rci
# rate.plot <- ggplot(ul.rci, aes(interval, distance)) +
#   geom_point() + 
#   facet_wrap(~mol) +
#   theme(strip.text.x = element_text(size = 7)) +
#   stat_smooth(method = "loess", se = F, size = 1) + 
#   ylab("Hellinger Distance") + 
#   xlab("Time Interval (Weeks)") +
#   theme_minimal()
# rate.plot
```

### How synchronous are community dynamics?
```{r}
alpha.ts <- zoo(alpha.div.active, env.data$date)
temp.ts <- zoo(env.data$temp, env.data$date)

require(synchrony)
alpha.sync.active <- community.sync(OTUs.active[,which(colMeans(OTUs.total.pa) >= 1)], nrands = 1000)
alpha.sync.total <- community.sync(OTUs.total[,which(colMeans(OTUs.total.pa) >= 1)], nrands = 1000)
alpha.sync.active
alpha.sync.total
```

### Explaining temporal beta diversity
```{r}
active.bd <- beta.div(OTUs.REL.active, nperm = 100, save.D = T, clock = T)
active.bd$SCBD[active.bd$SCBD ]
plot(env.data$temp, active.bd$LCBD)
total.bd <- beta.div(OTUs.REL.total, nperm = 100, save.D = T, clock = T)
plot(env.data$temp, total.bd$LCBD)
plot(total.bd$LCBD, type = "o")
total.bd$beta
active.bd$beta
plot(active.bd$SCBD, total.bd$SCBD)
abline(a = 0, b = 1)

active.comp <- beta.div.comp(OTUs.active, coef = "S", quant = T)
total.comp <- beta.div.comp(OTUs.total, coef = "S", quant = T)

samplingdays <- env.data$date
time.dist <- matrix(NA, nrow = length(samplingdays), ncol = length(samplingdays))
for(i in 1:length(samplingdays)){
  for(j in 1:length(samplingdays)){
    time.dist[i,j] <- (samplingdays[i] - samplingdays[j])
  }
}
time.dist <- as.dist(time.dist)

env.vars <- env.data[c("temp", "spc", "oxygen", "secchi", "ph", "tp", "tn", "doc")]
env.dist <- daisy(scale(env.vars), metric = "gower")

OTUs.pa <- decostand(OTUs.total, method = "pa")
OTUs.betapart <- beta.pair(x = OTUs.pa, index.family = "sorensen")
OTUs.nest <- OTUs.betapart$beta.sne
OTUs.turn <- OTUs.betapart$beta.sim
OTUs.sor <- OTUs.betapart$beta.sor
OTUs.bray.tot <- vegdist(OTUs.REL.total, method = "euclidean")
OTUs.bray.act <- vegdist(OTUs.REL.active, method = "euclidean")
OTUs.bray.sb <- seedbank.dist
OTUS.bray.trans <- transients.dist


timeddr <- data.frame(total_repl = as.vector(total.comp$repl),
                      total_rich = as.vector(total.comp$rich),
                      active_repl = as.vector(active.comp$repl),
                      active_rich = as.vector(active.comp$rich),
           EnvDist = as.vector(env.dist), 
           Time = as.vector(time.dist))
betapart.ddr.env <- timeddr %>% gather(key = Community, value = Dissimilarity, -EnvDist, -Time) %>% 
  ggplot(aes(x = EnvDist, y = Dissimilarity, color = Community)) + 
  geom_jitter(alpha = 0.1, size = 0.7) +
  geom_smooth(size = 2, method = "lm") +
  scale_y_continuous(limits = c(0,1))
betapart.ddr.time <- timeddr %>% gather(key = Community, value = Dissimilarity, -EnvDist, -Time) %>% 
  ggplot(aes(x = Time, y = Dissimilarity, color = Community)) + 
  geom_jitter(alpha = 0.1, size = 0.7) +
  geom_smooth(size = 2) +
  ylab("Community Dissimilarity\n") + 
  xlab("Time lag (days)") +
  scale_y_continuous(limits = c(0,1))
plot_grid(betapart.ddr.time + theme(legend.position = "none"), 
          betapart.ddr.env, ncol = 2, align = 'hv', axis = 'l',
          labels = c("A", "B")) +
  ggsave(filename = "figures/betapart_ddrs.pdf", width = 10, height = 6, units = "in", bg = "white")


# timeddr <- data.frame(Turnover = as.vector(OTUs.turn), 
#            Nestedness = as.vector(OTUs.nest),
#            Sorensen = as.vector(OTUs.sor),
#            Bray_tot = as.vector(OTUs.bray.tot),
#            Bray_act = as.vector(OTUs.bray.act),
#            Bray_sb = as.vector(OTUs.bray.sb),
#            EnvDist = as.vector(env.dist), 
#            Time = as.vector(time.dist))

# turn.time.mod <- (lm(Turnover ~ Time + I(Time^2) + I(Time^3) + I(Time^4) + I(Time^5), data = timeddr))
# summary(turn.time.mod)
# newdf <- data.frame(Time = min(timeddr$Time):max(timeddr$Time))
# preds <- predict(turn.time.mod, newdf)
# plot(Turnover ~ Time, timeddr, col = "gray70")
# lines(newdf$Time, preds, lwd = 2)
# 
# env.time.mod <- (lm(EnvDist ~ Time + I(Time^2) + I(Time^3) + I(Time^4) + I(Time^5), data = timeddr))
# summary(env.time.mod)
# newdf <- data.frame(Time = min(timeddr$Time):max(timeddr$Time))
# preds <- predict(env.time.mod, newdf)
# plot(EnvDist ~ Time, timeddr, col = "gray70")
# lines(newdf$Time, preds, lwd = 2)
# 
# plot(env.time.mod)
# 
# turn.env.mod <- lm(resid(turn.time.mod) ~ resid(env.time.mod))
# summary(turn.env.mod)
# plot(resid(env.time.mod), resid(turn.time.mod))
# 
# nest.time.mod <- (lm(Nestedness ~ Time, data = timeddr))
# summary(nest.time.mod)
# newdf <- data.frame(Time = min(timeddr$Time):max(timeddr$Time))
# preds <- predict(nest.time.mod, newdf)
# plot(Nestedness ~ Time, timeddr, col = "gray70")
# lines(newdf$Time, preds, lwd = 2)
# 
# nest.env.mod <- lm(resid(nest.time.mod) ~ resid(env.time.mod))
# summary(nest.env.mod)
# plot(resid(env.time.mod), resid(nest.time.mod))
  
# timeddr %>% gather(key = Partition, value = Beta, -EnvDist, -Time) %>% 
#   ggplot(aes(x = EnvDist, y = Bray, color = Partition)) + 
#   geom_point(alpha = 0.1) +
#   geom_smooth(size = 2)
# timeddr %>% gather(key = Partition, value = Beta, -EnvDist, -Time) %>% 
#   ggplot(aes(x = Time, y = Beta, color = Partition)) + 
#   geom_point(alpha = 0.1) +
#   geom_smooth(size = 2)

timeddr <- data.frame(Total = as.vector(OTUs.bray.tot),
           Active = as.vector(OTUs.bray.act),
           "Seed bank" = as.vector(OTUs.bray.sb),
           Transients = as.vector(OTUS.bray.trans),
           EnvDist = as.vector(env.dist), 
           Time = as.vector(time.dist))

ddr.env <- timeddr %>% select(-Seed.bank, -Transients) %>% 
  gather(key = Community, value = Dissimilarity, -EnvDist, -Time) %>% 
  ggplot(aes(x = EnvDist, y = (Dissimilarity), color = Community)) + 
  geom_point(alpha = 0.1, shape = 16, size = 0.7) +
  geom_smooth(size = 2) +
  #scale_y_log10() +
  labs(y = "Community Dissimilarity", x = "Environmental Dissimilarity") +
  scale_color_manual(values = my.cols[c(4,1)]) +
  scale_y_continuous(limits = c(0.25,1.5)) +
  ggsave("figures/dissimilarity-by-env.pdf", bg = "white", height = 8, width = 8)

ddr.time <- timeddr %>% select(-Seed.bank, -Transients) %>% 
  gather(key = Community, value = Dissimilarity, -EnvDist, -Time) %>% 
  ggplot(aes(x = Time, y = Dissimilarity, color = Community)) + 
  geom_vline(xintercept = 365*c(1,2), color = "grey", alpha = 0.5) +
  geom_jitter(alpha = 0.1, shape = 16, size = 0.7) +
  geom_smooth(size = 2) +
  ylab("Community Dissimilarity") + 
  xlab("Time lag (days)") +
  scale_color_manual(values = my.cols[c(4,1)]) +
  scale_y_continuous(limits = c(0.25,1.5)) +
  ggsave("figures/dissimilarity-by-time.pdf", bg = "white", height = 8, width = 8)

plot_grid(ddr.time + theme(legend.position = "none"), 
          ddr.env + theme(legend.position = "none"), ncol = 2, align = 'hv', axis = 'l',
          labels = c("A", "B")) +
  ggsave(filename = "figures/ddrs.pdf", width = 10, height = 6, units = "in", bg = "white")


#### Constrained Ordination
intervals <- as.vector(diff.Date(samplingdays))
w <- 1/(intervals/max(intervals))
aem.out <- adespatial::aem.time(n = length(samplingdays), w = w, moran = TRUE, plot.moran = T)
aem.eigs <- data.frame(aem.out$aem)
aem.out$values
colnames(aem.eigs) <- paste0("AEM",seq(1,ncol(aem.eigs)))

env.preds <- na.omit(env.data[c("sample.id", "temp", "spc", "oxygen", "secchi", "ph", "tp", "tn", "doc")])

OTUs.rel.act.subs <- OTUs.REL.active[design$sample.id[which(design$sample.type == "RNA")] %in% env.preds$sample.id,]
aem.subs <- aem.eigs[design$sample.id[which(design$sample.type == "RNA")] %in% env.preds$sample.id,]

# #active.varpar <- varpart(OTUs.rel.act.subs, env.preds)
active.rda.0 <- rda(OTUs.rel.act.subs ~ 1, data = aem.subs)
# active.rda.1 <- rda(OTUs.rel.act.subs ~ ., data = aem.subs)
# active.rda.2 <- rda(OTUs.rel.act.subs ~ AEM2 + AEM4 + AEM5 + AEM6, data = aem.subs)
# active.rda <- ordiR2step(active.rda.0, scope = formula(active.rda.2), direction = "forward")
# 
# active.rda.env.0 <- rda(OTUs.rel.act.subs ~ 1, data = env.preds)
# active.rda.env.1 <- rda(OTUs.rel.act.subs ~ ., data = env.preds)
# active.rda.env <- ordiR2step(active.rda.env.0, scope = formula(active.rda.env.1), direction = "forward")

envfit(active.rda.0, aem.subs)
time.mod <- model.matrix(~ -1 + AEM5 + AEM4 + AEM6 + AEM2, data = aem.subs)
env.mod <- model.matrix(~ temp + doc + oxygen + spc + ph, data = env.preds)
active.rda <- rda(OTUs.rel.act.subs ~ env.mod + Condition(time.mod))

active.rda.sum <- summary(active.rda)
env.vecs <- cbind.data.frame(center = 0, active.rda.sum$biplot[,c(1:2)])
rownames(env.vecs) <- c("Temp", "DOC", "DO", "SpC", "pH")
spec.scores <- cbind.data.frame(center = 0, scores(active.rda)$species)
site.scores <- as.data.frame(scores(active.rda)$sites)

active.site.scores <- site.scores %>% rownames_to_column(var = "sample.name") %>% inner_join(design)
active.spec.scores <- spec.scores[which(abs(spec.scores$RDA1) > 0.1 | abs(spec.scores$RDA2) > 0.1),] %>% 
  rownames_to_column(var = "OTU") %>% inner_join(OTUs.tax)

ggplot() + 
  # Add site scores as points
  # geom_point(data = site.scores, aes(x = RDA1, y = RDA2)) + 
  geom_point(data = active.site.scores, 
            aes(x = RDA1, y = RDA2, 
                color = month(date, label = T, abbr = T))) + 
  scale_color_viridis(discrete = T, "Month") +
  
  # Add species scores as red vectors
  geom_segment(data = active.spec.scores, 
               aes(x = center, y = center, xend = RDA1, yend = RDA2), 
               alpha = 0.5, color = "black",
             arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"), 
             show.legend = F) + 
  geom_label_repel(data = active.spec.scores, 
               aes(x = RDA1, y = RDA2), 
               label = active.spec.scores$Class, label.padding = .3, alpha = 0.7) + 

  # Add environmental vectors in blue
  geom_segment(data = env.vecs,
             aes(x = center, y = center, xend = RDA1, yend = RDA2),
             arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"),
             show.legend = F, alpha = .5, color = "blue") +
  geom_label_repel(data = env.vecs, aes(x = RDA1, y = RDA2),
                    label = rownames(env.vecs), color = "blue", 
                   segment.alpha = 0.3, alpha = 0.7) + 
  theme_cowplot() + 
  coord_fixed() + 
  ggsave("figures/rda_active_aem.pdf", bg = "white", height = 8, width = 8)

matplot(aem.subs[,c(5,4,6,2)], type = 'l')

OTUs.rel.tot.subs <- OTUs.REL.total[design$sample.id[which(design$sample.type == "DNA")] %in% env.preds$sample.id,]
aem.subs <- aem.eigs[design$sample.id[which(design$sample.type == "DNA")] %in% env.preds$sample.id,]

#active.varpar <- varpart(OTUs.rel.act.subs, env.preds)
total.rda.0 <- rda(OTUs.rel.tot.subs ~ 1, data = aem.subs)
envfit(total.rda.0, aem.subs)


#total.rda.1 <- rda(OTUs.rel.tot.subs ~ ., data = aem.subs)
# total.rda.2 <- rda(OTUs.rel.tot.subs ~ AEM1 + AEM2 + AEM4 + AEM5 + AEM6 + AEM10, data = aem.subs)
# #total.rda <- ordiR2step(active.rda.0, scope = formula(active.rda.2), direction = "forward")
# total.rda.env.0 <- rda(OTUs.rel.tot.subs ~ 1, data = env.preds)
# total.rda.env.1 <- rda(OTUs.rel.tot.subs ~ ., data = env.preds)
# total.rda.env <- ordiR2step(total.rda.env.0, scope = formula(total.rda.env.1), direction = "forward")
time.mod <- model.matrix(~ -1 + AEM1 + AEM2 + AEM4 + AEM5 + AEM6 + AEM10, data = aem.subs)
env.mod <- model.matrix(~ temp + doc + oxygen + spc + ph, data = env.preds)
total.rda <- rda(OTUs.rel.tot.subs ~ env.mod + Condition(time.mod))

total.rda.sum <- summary(total.rda)
env.vecs <- cbind.data.frame(center = 0, total.rda.sum$biplot[,c(1:2)])
rownames(env.vecs) <- c("Temp", "DOC", "DO", "SpC", "pH")
spec.scores <- cbind.data.frame(center = 0, scores(total.rda)$species)
site.scores <- as.data.frame(scores(total.rda)$sites)

total.site.scores <- site.scores %>% rownames_to_column(var = "sample.name") %>% inner_join(design)
total.spec.scores <- spec.scores[which(abs(spec.scores$RDA1) > 0.1 | abs(spec.scores$RDA2) > 0.1),] %>% 
  rownames_to_column(var = "OTU") %>% inner_join(OTUs.tax)

ggplot() + 
  # Add site scores as points
  # geom_point(data = site.scores, aes(x = RDA1, y = RDA2)) + 
  geom_point(data = total.site.scores, 
            aes(x = RDA1, y = RDA2, 
                color = month(date, label = T, abbr = T))) + 
  scale_color_viridis(discrete = T, "Month") +
  
  # Add species scores as red vectors
  geom_segment(data = total.spec.scores, 
               aes(x = center, y = center, xend = RDA1, yend = RDA2), 
               alpha = 0.5, color = "black",
             arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"), 
             show.legend = F) + 
  geom_label_repel(data = total.spec.scores, 
               aes(x = RDA1, y = RDA2), 
               label = total.spec.scores$Class, label.padding = 0.3, alpha = 0.7) + 

  # Add environmental vectors in blue
  geom_segment(data = env.vecs,
             aes(x = center, y = center, xend = RDA1, yend = RDA2),
             arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"),
             show.legend = F, alpha = .5, color = "blue") +
  geom_label_repel(data = env.vecs, aes(x = RDA1, y = RDA2),
                    label = rownames(env.vecs), color = "blue", 
                   segment.alpha = 0.3, alpha = 0.7) + 
  theme_cowplot() + 
  coord_fixed() + 
  ggsave("figures/rda_total_aem.pdf", bg = "white", height = 7, width = 7)

aem.subs[,c(1,2,4,5,6,10)] %>% rownames_to_column(var = "Time") %>% gather(AEM, Evec, -Time) %>% 
  ggplot(aes(x = as.numeric(Time), y = Evec, color = AEM)) +
  geom_line()

```

# Is negative frequency dependence disproportionately stronger on rare than common taxa?
```{r}
# calc.neg.freq.dep <- function(sbs){
#   y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
#   for(i in 1:(nrow(sbs)-1)){
#     y.s[i,] = log(sbs[i+1,] / sbs[i,])
#   }
#   eq.freq <- vector(length = ncol(y.s))
#   fd <- vector(length = ncol(y.s))
#   sbs.rel <- decostand(sbs[-nrow(sbs),], method = "total")
#   for(i in 1:ncol(y.s)){
#     mod <- lm(y.s[,i] ~ sbs.rel[,i])
#     eq.freq[i] <- -coef(mod)[1] / coef(mod)[2]
#     fd[i] <- coef(mod)[2]
#   }
#   # plot(log10(eq.freq), log10(-fd), 
#   #      ylab = "log10 Negative Freq. Dependence",
#   #      xlab = "log10 Equilibrium Frequency")
#   return(na.omit(cbind.data.frame(eq.freq, fd)))
# }

# using mean rel abund
calc.neg.freq.dep <- function(sbs){
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  eq.freq <- vector(length = ncol(y.s))
  fd <- vector(length = ncol(y.s))
  sbs.rel <- decostand(sbs[-nrow(sbs),], method = "total")
  for(i in 1:ncol(y.s)){
    mod <- lm(y.s[,i] ~ sbs.rel[,i])
    eq.freq[i] <- mean(sbs.rel[,i])
    fd[i] <- coef(mod)[2]
  }
  # plot(log10(eq.freq), log10(-fd), 
  #      ylab = "log10 Negative Freq. Dependence",
  #      xlab = "log10 Equilibrium Frequency")
  return(na.omit(cbind.data.frame(eq.freq, fd)))
}

shuffle.ts <- function(species.ts){
  species.ts[sample(1:length(species.ts))]
}

get.null.distr <- function(comm, iters = 5000){
  pb <- progress_bar$new(total = iters)
  nfd.cov <- rep(NA, iters)
  for(i in 1:iters){
    nullcom <- apply(comm, MARGIN = 2, FUN = shuffle.ts) # randomize sampling order
    nfd <- calc.neg.freq.dep(nullcom)
    nfd <- nfd %>% filter(-fd > 0)
    nfd.cov[i] <- cov(log(nfd$eq.freq), log(-nfd$fd))
    pb$tick()
  }
  return(nfd.cov)
}

# OTUs.ranked <- OTUs.total[,names(sort(colSums(OTUs.total), decreasing = T))]

# separate OTUs present across the nearly whole time series
OTUs.total.persistent <- OTUs.total[,which(
  colMeans(decostand(OTUs, method = "pa")) >= .8)]
OTUs.active.persistent <- OTUs.active[,colnames(OTUs.total.persistent)]

# How many taxa to consider?
ncol(OTUs.total.persistent)


OTUs.total.nozero <- replace(OTUs.total.persistent, which(OTUs.total.persistent == 0), .5)
nfd.total <- calc.neg.freq.dep(OTUs.total.nozero)
nfd.total %>% filter(-fd > 0) -> nfd.total
nfd.total.cov <- cov(log(nfd.total$eq.freq), log(-nfd.total$fd))
nfd.total %>% ggplot(aes(x = eq.freq, y = -fd)) +
  geom_point() + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm") + 
  ylab("Negative Frequency Dependence") + 
  xlab("Equilibrium Frequency")

OTUs.active.nozero <- replace(OTUs.active.persistent, which(OTUs.active.persistent == 0), .5)
nfd.active <- calc.neg.freq.dep(OTUs.active.nozero)
nfd.active %>% filter(-fd > 0) -> nfd.active
nfd.active.cov <- cov(log(nfd.active$eq.freq), log(-nfd.active$fd))
nfd.active %>% ggplot(aes(x = eq.freq, y = -fd)) +
  geom_point() + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm") + 
  ylab("Negative Frequency Dependence") + 
  xlab("Equilibrium Frequency")


# Generate null distributions
iternum <- 1000
tot.nd <- get.null.distr(comm = OTUs.total.nozero, iters = iternum)
act.nd <- get.null.distr(comm = OTUs.active.nozero, iters = iternum)

# Compare observed with null
hist(tot.nd, breaks = 20); abline(v = nfd.total.cov, col = "red")
(tot.nfd.pval <- rank(c(nfd.total.cov, tot.nd))[1] / length(c(nfd.total.cov, tot.nd)))
hist(act.nd, breaks = 20); abline(v = nfd.active.cov, col = "red")
(act.nfd.pval <- rank(c(nfd.active.cov, act.nd))[1] / length(c(nfd.active.cov, act.nd)))
# hist(sb.nd, breaks = 20); abline(v = nfd.sb.cov, col = "red")
# (sb.nfd.pval <- rank(c(nfd.sb.cov, sb.nd))[1] / length(c(nfd.sb.cov, sb.nd)))

# nfd.obs <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = nfd.total.cov),
#                  cbind.data.frame(subset = "Active", nfd = nfd.active.cov),
#                  cbind.data.frame(subset = "Seed Bank", nfd = nfd.sb.cov)) 
# 
# nfd.nulldists <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = tot.nd),
#                  cbind.data.frame(subset = "Active", nfd = act.nd),
#                  cbind.data.frame(subset = "Seed Bank", nfd = sb.nd)) 

nfd.obs <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = nfd.total.cov),
                 cbind.data.frame(subset = "Active", nfd = nfd.active.cov)) 

nfd.nulldists <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = tot.nd),
                 cbind.data.frame(subset = "Active", nfd = act.nd)) 

nfd.nulldists %>% 
  ggplot(aes(y = nfd, x = subset)) + 
  #geom_jitter(color = "grey", alpha = 0.1) + 
  geom_violin(alpha = 0.5, show.legend = F, fill = "grey") + 
  geom_point(data = nfd.obs, aes(x = subset, y = nfd), 
             color = "black", size = 5, show.legend = F) + 
  scale_fill_manual(values = my.cols[c(1,4)]) +
  coord_flip() +
  ylab("Covariance between Equilibrium Frequency\n and Negative Frequency Dependence") + 
  xlab("") + 
  ggsave("figures/negative_freq_depend.pdf", bg = "white", width = 6, height = 4, units = "in")

# Calculate standardized effect sizes
nfd.summary <- nfd.nulldists %>% 
  group_by(subset) %>% summarize(mean.nfd = mean(nfd), sd.nfd = sd(nfd)) %>% 
  left_join(nfd.obs) %>% mutate(nfd.ses = (nfd - mean.nfd)/sd.nfd) %>% 
  mutate(nfd.strength = nfd / mean.nfd)
nfd.summary
```

```{r}
# What is the effect of taxa number on response
nfd.response <- function(OTUs = NULL, OTUs.active = NULL, num.seq = 10, num.iter = 999, plotit = F, incidence.freq = 0.9, max.taxa = NULL){
  
  # separate OTUs present across the whole time series
  OTUs.ranked <- OTUs[,names(sort(colSums(OTUs), decreasing = T))]
  OTUs.persistent <- OTUs.ranked[,which(
    colMeans(decostand(OTUs.ranked, method = "pa")) >= incidence.freq)]
  
  # If we're looking at Active OTUs, make the switch here
  if(!is.null(OTUs.active)){
    if(is.null(OTUs)) print("please supply total otu table")
    OTUs.persistent <- OTUs.active[,colnames(OTUs.persistent)]
  }
  
  # decide how many values of cutoffs to test
  max.taxa <- floor(seq(3, min(max.taxa, ncol(OTUs.persistent)), length.out = num.seq))
  output <- data.frame(taxa = max.taxa)
  
  # loop through different otu thresholds
  for(i in 1:length(max.taxa)){
    if(max.taxa[i] > ncol(OTUs.persistent)) return(output)
    otu.thresh <- min(max.taxa[i], ncol(OTUs.persistent))
    OTUs.nozero <- OTUs.persistent[,1:otu.thresh] + 10^-10
    nfd <- calc.neg.freq.dep(OTUs.nozero)
    nfd <- nfd %>% filter(eq.freq > 10^-9) %>% filter(-fd > 0)
    nfd.cov <- cov(log(nfd$eq.freq), log(-nfd$fd))
    
    if(plotit){
      nfd %>% 
        ggplot(aes(x = eq.freq, y = -fd)) +
        geom_point() + 
        scale_y_log10() + 
        scale_x_log10() +
        geom_smooth(method = "lm") + 
        ylab("Negative Frequency Dependence") + 
        xlab("Equilibrium Frequency")
    }
    
    nd <- get.null.distr(comm = OTUs.nozero, iters = num.iter)
    nfd.ses <- (nfd.cov - mean(nd)/sd(nd))
    nfd.pval <- rank(c(nfd.cov, nd))[1] / length(c(nfd.cov, nd))
    cov.strength <- nfd.cov / mean(nd)
    se <- sd(nd)/sqrt(num.iter)
    output[i,"p"] <- nfd.pval
    output[i,"ses"] <- nfd.ses
    output[i,"strength"] <- cov.strength
    output[i,"stderr"] <- se
    output[i,"mean"] <- mean(nd)
    output[i,"nfdcov"] <- nfd.cov
  }
  
  return(output)
}


active.nfd.response <- nfd.response(OTUs = OTUs.total, OTUs.active = OTUs.active, num.seq = 30, num.iter = 999, max.taxa = 100)
total.nfd.response <- nfd.response(OTUs = OTUs.total, OTUs.active = NULL, num.seq = 30, num.iter = 999, max.taxa = 100)
saveRDS(active.nfd.response, file = "analysis/temp-results/nfd_response_active.rda")
saveRDS(total.nfd.response, file = "analysis/temp-results/nfd_response_total.rda")
active.nfd.response <- readRDS(file = "analysis/temp-results/nfd_response_active.rda")
total.nfd.response <- readRDS(file = "analysis/temp-results/nfd_response_total.rda")

nfd.response.df <- rbind.data.frame(
  cbind.data.frame(subset = "active", active.nfd.response),
  cbind.data.frame(subset = "total", total.nfd.response))
nfd.response.df %>% 
#  filter(p < 0.05) %>% 
  ggplot(aes(x = taxa, y = strength, color = subset)) +
  geom_point() +
  geom_smooth()

cbind.data.frame(taxa = active.nfd.response$taxa, 
      strength = active.nfd.response$strength / total.nfd.response$strength) %>% 
  ggplot(aes(x = taxa, y = strength)) +
  geom_point() +
  geom_abline(slope = 0, intercept = 1, color = "gray") +
  geom_smooth(method = "loess")
```

# Reappeareance Scores
```{r}
# Calculate a species 

reappearance.score <- function(DNA, RNA){
  # calculate # of times present but not active
  otu.diff <- decostand((DNA + RNA), method = "pa") - decostand(RNA, method = "pa")
  otu.diff.weighted <- otu.diff # * decostand(DNA, method = "total")
  return(colSums(otu.diff.weighted))
}

calc.score <- function(species){
  length()
}

(OTUs.total - OTUs.active)[,1]
reappearance.score <- function(DNA, RNA){
  
  otu.diff <- decostand((DNA + RNA), method = "pa") - 
    decostand(RNA, method = "pa")
  otuscores <- vector(length = ncol(otu.diff))
  
  for(j in 1:ncol(otu.diff)){
    otuscore <- 0
    for(i in 1:(nrow(otu.diff)-1)){
      if(length(unique(otu.diff[c(i,i+1), j])) > 1){
        otuscore <- otuscore + 1
      }
    }
    otuscores[j] <- otuscore
    print(paste("completed otu #", j))
  }
  
  return(otuscores)
}

otu.reap <- reappearance.score(DNA = OTUs.total, RNA = OTUs.active)
otu.pres <- colSums(decostand((OTUs.total + OTUs.active), method = "pa"))
plot(otu.pres, otu.reap/otu.pres, ylab = "# of times present in seed bank but not active",
     xlab = "# of times present in lake")

cbind.data.frame(times_inactive = otu.reap, presences = otu.pres) %>%
  mutate(prop_inactive = times_inactive/presences, 
         incidence = presences / (length(env.data$date) - 1)) %>% 
  ggplot(aes(y = prop_inactive, x = incidence)) + 
  geom_jitter(width = 0.005, height = 0.005, alpha = 0.1) + 
  geom_smooth() + 
  ylab("Inactive Frequency") +
  xlab("Incidence in Total Community") + 
  ggsave("figures/incidence_by_inactivity.pdf", bg = "white", width = 6, height = 6)

subset(OTUs.tax, OTU %in% names(sort(otu.reap, decreasing = T)[1:20]))

# pull out taxa who are present > 50% of time but inactive > 50% of time
# these may be potential seed bankers or sink populations form soils
cbind.data.frame(times_inactive = otu.reap, presences = otu.pres) %>%
  rownames_to_column(var = "OTU") %>% 
  mutate(prop_inactive = times_inactive/presences, 
         incidence = presences / (length(env.data$date) - 1)) %>% 
  filter(incidence > 0.5, prop_inactive > 0.5) -> seedbank.taxa
subset(OTUs.tax, OTU %in% seedbank.taxa$OTU)
```

# Taxa Time Relationship
```{r}
ttr.calc <- function(comm){
  comm.pa <- decostand(comm, method = 'pa')
  weeknum <- 1:nrow(comm.pa)
  otusum <- numeric(nrow(comm.pa))
  for(week in weeknum){
    otusum[week] <- ifelse(week > 1, sum(colSums(comm.pa[1:week,]) > 0), 
                           sum(comm.pa[1,]))
  }
  return(cbind.data.frame(week = weeknum, taxa = otusum))
}

active.ttr <- ttr.calc(OTUs.active)
total.ttr <- ttr.calc(OTUs.total)
transient.ttr <- ttr.calc(transients)

ttrs <- rbind.data.frame(cbind.data.frame(subset = "active", active.ttr),
                 cbind.data.frame(subset = "total", total.ttr),
                 cbind.data.frame(subset = "transients", transient.ttr))
ttrs %>% 
  ggplot(aes(x = week, y = taxa, color = subset)) + 
  geom_line(size = 2) + 
  ggsave("figures/taxa_time_relationships.pdf", bg = "white", height = 8, width = 8)
```

# Conditionally Rare Taxa
```{r}
identifyCRT <- function(spec.dyn){
  n <- length(spec.dyn)
  skew.numer <- sum((spec.dyn - mean(spec.dyn))^3)/n
  skew.denom <- (sum((spec.dyn - mean(spec.dyn))^2)/n)^(3/2)
  kurt.num <- sum((spec.dyn - mean(spec.dyn))^4)/n
  kurt.denom <- (sum((spec.dyn - mean(spec.dyn))^2)/n)^2
  skewness <- skew.numer / skew.denom
  kurtosis <- kurt.num / kurt.denom
  b <- (1 + skewness^2) / (kurtosis + 3)
  return(b)
 }

activeCRTs <- apply(X = OTUs.REL.active[,which(colMeans(OTUs.active.pa) > 0.5)], MARGIN = 2, FUN = identifyCRT)
totalCRTs <- apply(X = OTUs.REL.total[,which(colMeans(OTUs.total.pa) > 0.5)], MARGIN = 2, FUN = identifyCRT)
hist(activeCRTs)
hist(totalCRTs)


candidateCRTs <- names(activeCRTs[which(activeCRTs > 0.9)])
candidateCRTs <- OTUs.REL.active[,candidateCRTs]
candidateCRTs <- candidateCRTs[,apply(candidateCRTs, 2, max) > 0.01]
matplot(candidateCRTs, type = 'l')
subset(OTUs.tax, OTU %in% colnames(candidateCRTs))


```


```{r}
# shuffle.ts <- function(species.ts){
#   species.ts[sample(1:length(species.ts))]
# }
# 
# get.null.mrs <- function(comm, iters = 100){
#   pb <- progress_bar$new(total = iters)
#   mrs.null <- rep(NA, iters)
#   for(i in 1:iters){
#     nullcom <- apply(comm, MARGIN = 2, FUN = shuffle.ts) # randomize sampling order
#     nullcom.ranks <- decostand(nullcom, method = "rank")
#     S <- specnumber(nullcom.ranks[-nrow(nullcom.ranks),])
#     nmrs <- rowSums(abs(diff(nullcom.ranks))) / S
#     mrs.null[i] <- mean(nmrs)
#     pb$tick()
#   }
#   return(mrs.null)
# }
# 
# total.null <- get.null.mrs(OTUs.total)
# active.null <- get.null.mrs(OTUs.active)
# 
# OTUs.total.ranks <- decostand(OTUs.total, method = "rank")
# MRS.total <- rowSums(abs(diff(OTUs.total.ranks))) / specnumber(OTUs.total.ranks[-nrow(OTUs.total.ranks),])
# plot(MRS.total, type = 'l')
# 
# OTUs.active.ranks <- decostand(OTUs.active, method = "rank")
# MRS.active <- rowSums(abs(diff(OTUs.active.ranks))) / specnumber(OTUs.active.ranks[-nrow(OTUs.active.ranks),])
# points(MRS.active, type = 'l', col = 'red')
# 
# MRS.total.ses <- (MRS.total - mean(total.null)) / sd(total.null)
# MRS.active.ses <- (MRS.active - mean(active.null)) / sd(active.null)
# matplot(cbind(MRS.total.ses, MRS.active.ses), type = 'l')
# 
# plot(env.vars$temp[-123], MRS.total)
# plot(env.vars$temp[-123], MRS.active)
# 
# plot(MRS.total, MRS.active)
# abline(0,1)
```

```{r}
maxgrowth <- function(sbs){
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  max.growth.time <- apply(y.s, 2, which.max)
  max.growths <- apply(y.s, 2, max)
  return(data.frame(
    rate = max.growths, sample = max.growth.time))
}

env.vars <- env.data[c("temp", "spc", "oxygen", "secchi", "ph", "tp", "tn", "doc")]
env.dist <- daisy(scale(env.vars), metric = "gower")
env.pca <- prcomp(na.omit(env.vars), scale. = T)

env.pcs <- as.data.frame(env.pca$x) %>% rownames_to_column(var = "sample.id") 
env.pcs$sample.id <- as.numeric(env.pcs$sample.id)
env.pcs <- env.pcs %>% right_join(env.data)

OTUs.total.persistent <- OTUs.total[,which(colMeans(OTUs.total.pa) > 0.9)]
OTUs.active.persistent <- OTUs.active[,colnames(OTUs.total.persistent)]
active.max.growths <- maxgrowth(OTUs.active.persistent + 1)
active.max.growths <- active.max.growths %>% mutate(sample.id = sample)

env.pcs.subset <- subset(env.pcs, sample.id %in% active.max.growths$sample.id) %>% inner_join(active.max.growths)  

ggplot() +
  geom_jitter(data = env.pcs.subset, 
              aes(x = PC1, y = PC2, size = rate, color = month(date, label = T, abbr = T)),
              alpha = 0.5) +
  geom_segment(data = as.data.frame(env.pca$rotation),
               aes(x = 0, y = 0, xend = PC1, yend = PC2),
               arrow = arrow(length = unit(.3, "cm")),
               show.legend = F, alpha = 1, color = "blue") +
  geom_label_repel(data = as.data.frame(env.pca$rotation), 
            aes(x = PC1, y = PC2, label = rownames(env.pca$rotation)),
            box.padding = .5, point.padding = .5, alpha = 0.9) +
  scale_color_viridis(discrete = T) +
  coord_fixed() +
  ggsave("figures/max_growth_envs.pdf", bg = "white", width = 8, height = 8)
```

