---
title: "University Lake Microbial Seed Bank"
author: "Nathan Wisnoski, Jay Lennon"
date: "1/19/2018"
output: html_document
---

```{r setup, include=FALSE}
# Load packages

require(vegan)
require(tidyverse)
require(lubridate)
require(viridis)
require(betapart)
require(cowplot)
#require(xts)
require(progress)
#require(cluster)
require(adespatial)
require(ggrepel)
require(colorspace)
require(parallel)
library(foreach)
library(doParallel)
library(forecast)
library(broom)
source("bin/mothur_tools.R")

cv <- function(x){
  sd(x) / abs(mean(x))
}

shifter <- function(x, n = 1) {
  if (n == 0) x else c(tail(x, -n), head(x, n))
}
```

Next, we'll set the aesthetics of the figures we will produce. 
```{r figure_setup}
#my.cols <- RColorBrewer::brewer.pal(n = 4, name = "Greys")[3:4]

# Set theme for figures in the paper
theme_set(theme_classic() + 
  theme(axis.title = element_text(size = 16),
        axis.title.x = element_text(margin = margin(t = 15, b = 15)),
        axis.title.y = element_text(margin = margin(l = 15, r = 15)),
        axis.text = element_text(size = 14),
        axis.text.x = element_text(margin = margin(t = 5)),
        axis.text.y = element_text(margin = margin(r = 5)),
        #axis.line.x = element_line(size = 1),
        #axis.line.y = element_line(size = 1),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_line(size = 1),
        axis.ticks.y = element_line(size = 1),
        axis.ticks.length = unit(.1, "in"),
        panel.border = element_rect(color = "black", fill = NA, size = 1.5),
        legend.title = element_blank(),
        legend.text = element_text(size = 14),
        strip.text = element_text(size = 14),
        strip.background = element_blank()
        ))
```

## Import Data
Here, we read in the processed sequence files from mothur (shared and taxonomy) and a design of the sampling. We also load in the environmental data. We then remove the mock community from the dataset and ensure the the design and OTU table are aligned by row.
```{r}
# Define Inputs
# Design = general design file for experiment
# shared = OTU table from mothur with sequence similarity clustering
# Taxonomy = Taxonomic information for each OTU
design <- "data/UL_timeseries_design.csv"
shared <- "data/ul_resgrad.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.pick.opti_mcc.shared"
taxon  <- "data/ul_resgrad.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.pick.opti_mcc.0.03.cons.taxonomy"

# Import Design
design <- read_csv(file = design)

# Import Shared Files
OTUs <- read.otu(shared = shared, cutoff = "0.03")    # 97% Similarity

# Import Taxonomy
OTU.tax <- read.tax(taxonomy = taxon, format = "rdp")


# Load environmental data
env.data <- read.table("data/ul-seedbank.env.txt", sep="\t", header=TRUE)
env.data$date <- as.Date(parse_date_time(env.data$date, "m d y"))
env.data$doc[which(env.data$doc == "**")] <- NA
env.data$doc <- as.numeric(as.character(env.data$doc))
summary(env.data)
design <- left_join(design, env.data[,c("sample.id", "date")])
env.data <- env.data[-which(!(env.data$date %in% design$date)),]

# Subset to just the UL timeseries samples
OTUs <- OTUs[str_which(rownames(OTUs), "UL"),]

# make sure OTU table matches up with design order
OTUs <- OTUs[match(design$sample.name, rownames(OTUs)),]
```

# envplot
```{r}
env.data %>% select(date, temp, spc, secchi, ph, chla, tp, tn, doc) %>% 
  mutate(tn = log10(tn), tp = log10(tp), chla = log10(chla)) %>% 
  rename("Log10[Chl a (µg/L)]" = chla,
         "Dissolved organic carbon (mg/L)" = doc,
         #"Dissolved oxygen (mg/L)" = oxygen,
         "pH" = ph,
         "Secchi depth (m)" = secchi,
         "Specific conductivity (µS)" = spc,
         "Temperature (°C)" = temp,
         "Log10[Total nitrogen (mg/L)]" = tn,
         "Log10[Total phosphorus (µg/L)]" = tp) %>% 
  gather(-date, key = "variable", value = "val") %>% 
  ggplot(aes(x = date, y = val)) + 
  geom_point(alpha = 0.5) + 
  geom_line() +
  scale_x_date(date_labels = "%b\n%Y", date_breaks = "6 months") +
  labs(x = "", y = "") +
  facet_wrap(~variable, scales = "free_y", ncol = 2) + 
  ggsave("figures/env_data.pdf", width = 8, height = 8)
```

# transform OTUs
```{r data format, include=FALSE}

# Make rel abund matrices and split into active total comms
# OTUs <- OTUs[,-which(colSums(OTUs) < 5)]
set.seed(47405)
OTUs <- rrarefy(OTUs, sample = min(rowSums(OTUs)))

OTUs.active <- OTUs[which(design$sample.type == "RNA"),]
OTUs.total <- OTUs[which(design$sample.type == "DNA"),]
OTUs.total <- OTUs.total + decostand(OTUs.active, method = "pa")
OTUs.total.pa <- decostand(OTUs.total, method = "pa")
OTUs.active.pa <- decostand(OTUs.active, method = "pa")

#OTUs.REL <- decostand(OTUs, method = "total")
OTUs.REL <- decostand(OTUs, method = "hellinger")
OTUs.REL.total <- OTUs.REL[which(design$sample.type == "DNA"),]
OTUs.REL.active <- OTUs.REL[which(design$sample.type == "RNA"),]

# separate OTUs present across the nearly whole time series
OTUs.total.persistent <- OTUs.total[,which(
  colMeans(decostand(OTUs.total, method = "pa")) >= .8)]
OTUs.active.persistent <- OTUs.active[,colnames(OTUs.total.persistent)]


# Define color pallette
my.cols <- viridis(4, begin = .3, end = 0.8)
```

# Characterize temporal dynamics of the bacterioplankton community 

## How does alpha diversity vary over time?
```{r}
alpha.div.total <- rowSums(OTUs.total.pa) #exp(diversity(OTUs.total, index = "shannon")) # spec equivs
alpha.div.active <- rowSums(OTUs.active.pa) #exp(diversity(OTUs.active, index = "shannon"))
data.frame(date = env.data$date, Total = alpha.div.total, Active = alpha.div.active) %>% 
  mutate(diversity_ratio = Total / Active) %>% 
  #gather(community, diversity, -date) %>% 
  ggplot(aes(x = date, y = diversity_ratio)) + 
  # geom_vline(xintercept = as.numeric(as.Date(c("2014-01-01", "2015-01-01"))),
  #            linetype = 2, color = "grey", alpha = 0.8) +
  # geom_rect(mapping = aes(xmin = as_date("2013-04-01"), xmax = as_date("2013-07-31"),
  #                         ymin = min(diversity_ratio), ymax = max(diversity_ratio)),
  #           fill = "gray90", alpha = 0.1) +
  # geom_rect(mapping = aes(xmin = as_date("2014-04-01"), xmax = as_date("2014-07-31"),
  #                         ymin = min(diversity_ratio), ymax = max(diversity_ratio)),
  #           fill = "gray90", alpha = 0.1) +
  # geom_rect(mapping = aes(xmin = as_date("2015-04-01"), xmax = max(date),
  #                         ymin = min(diversity_ratio), ymax = max(diversity_ratio)),
  #           fill = "gray90", alpha = 0.1) + 
  geom_rect(mapping = aes(xmin = as_date("2013-10-01"), xmax = as_date("2014-03-31"),
                          ymin = min(diversity_ratio), ymax = max(diversity_ratio)),
            fill = "gray90", alpha = 0.1) +
  geom_rect(mapping = aes(xmin = as_date("2014-10-01"), xmax = as_date("2015-03-31"),
                          ymin = min(diversity_ratio), ymax = max(diversity_ratio)),
            fill = "gray90", alpha = 0.1) +
  geom_line(size = .7, alpha = .5, show.legend = F) +
  geom_point(aes(color = lubridate::month(date, label = T)), size = 3, alpha = 0.8) +
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) +
  geom_smooth(alpha = 0.5, color = "black", span = .5) +
  # ylab(expression(paste(alpha,"-diversity ratio (total / active richness)"))) +
  ylab("Seed bank richness ratio \n(total : active)") +
  xlab("Time") +
  scale_x_date(date_breaks = "3 month", 
               date_minor_breaks = "1 month", 
               date_labels = "%b") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none") +
  #scale_color_manual("Community Subset", values = c("#b2df8a", "#1f78b4")) +
  #scale_color_manual(values = my.cols[c(4,1)]) +
  #scale_y_log10() +
  #annotation_logticks(sides = "l") +
  ggsave("figures/fig4_nomod.pdf", bg = "white", height = 3/4*6, width = 6) +
  ggsave("figures/fig4_nomod.png", height = 3/4*6, width = 6, dpi = 600)

alpha_dat <- data.frame(date = env.data$date, sample.id = env.data$sample.id, Total = alpha.div.total, Active = alpha.div.active) %>% 
  mutate(diversity_ratio = Total / Active)
alpha_ts <- arima(alpha_dat$diversity_ratio,
                  order = c(1, 0, 0))
alpha_ts

env.preds <- (env.data[c("sample.id", "temp", "secchi")])
env.tsmod <- na.omit(env.preds) %>% inner_join(alpha_dat, by = "sample.id")
env_preds_scaled <- model.matrix(~ temp + secchi, data = as.data.frame(scale(env.tsmod[,c(2:3)])))[,-1]

# alpha_env_ts <- arima(env.tsmod$diversity_ratio, 
#       order = c(1,0,0), 
#       xreg = env_preds_scaled[,])

alpha_env_ts <- auto.arima(y = env.tsmod$diversity_ratio, 
                           xreg = env_preds_scaled)
alpha_env_ts_pred <- forecast(alpha_env_ts, level = c(95), xreg = env_preds_scaled)

forecast::autoplot(alpha_env_ts, 
                   ts.colour = "black", 
                   predict.colour = "red", predict.linetype = "dashed",
                   conf.int.fill = "red")

plot(alpha_env_ts$residuals)
plot(fitted(alpha_env_ts))

model.fit.to.plot <- data.frame(
  "date" = env.tsmod$date,
  "fitted" = alpha_env_ts_pred$fitted,
  "upper" = alpha_env_ts_pred$upper,
  "lower" = alpha_env_ts_pred$lower
)
names(model.fit.to.plot)[c(3,4)] = c("upper", "lower")

ggplot(data = env.tsmod) + 
  geom_rect(mapping = aes(xmin = as_date("2013-10-01"), xmax = as_date("2014-03-31"),
                          ymin = min(diversity_ratio), ymax = max(diversity_ratio)),
            fill = "white", alpha = 0.1) +
  geom_rect(mapping = aes(xmin = as_date("2014-10-01"), xmax = as_date("2015-03-31"),
                          ymin = min(diversity_ratio), ymax = max(diversity_ratio)),
            fill = "white", alpha = 0.1) +
  
  # add ARIMA model
  
  geom_line(data = model.fit.to.plot, aes(x = date, y = fitted), size = .8) +
  geom_ribbon(data = model.fit.to.plot, aes(x = date, ymin = lower, ymax = upper), fill = "grey50", alpha = 0.2) + 
  
  # add data points
  #geom_line(aes(x = date, y = diversity_ratio), size = .7, alpha = .25, show.legend = F) +
  geom_point(aes(x = date, y = diversity_ratio, color = lubridate::month(date, label = T)), size = 3, alpha = 0.8) +
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) +
  
  
  # ylab(expression(paste(alpha,"-diversity ratio (total / active richness)"))) +
  ylab("Seed bank richness ratio \n(total : active)") +
  xlab("Time") +
  scale_x_date(date_breaks = "3 month", 
               date_minor_breaks = "1 month", 
               date_labels = "%b") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1),
        legend.position = "none") +
  ggsave("figures/fig4.pdf", bg = "white", height = 3/4*6, width = 6) +
  ggsave("figures/fig4.png", width = 6, height = 3/4*6, dpi = 1000)

```


## How does community composition change over time?
### Ordination methods (abundance-based)
```{r}

active.pca <- rda(OTUs.REL.active)

# 
# OTUs.dist <- vegdist(OTUs.REL, method = "euclidean")
# OTUs.pcoa <- cmdscale(OTUs.dist, eig = T)
var1 <- round(eigenvals(active.pca)[1] / sum(eigenvals(active.pca)) * 100, 2)
var2 <- round(eigenvals(active.pca)[2] / sum(eigenvals(active.pca)) * 100, 2)

# OTUs.traj <- OTUs.pcoa$points
# colnames(OTUs.traj) <- c("Comp.1", "Comp.2")
# active.traj <- OTUs.traj[which(design$sample.type == "RNA"),]
# total.traj <- OTUs.traj[which(design$sample.type == "DNA"),]

active.traj.df <- cbind.data.frame(
  sample.id = design$sample.id[which(design$sample.type == "RNA")], 
  as.data.frame(scores(active.pca, display = "sites"))) 

active.spec.scores.pc <- as.data.frame(scores(active.pca, display = "species")) %>% rownames_to_column(var = "OTU") %>% 
  filter(PC1 > .2 | PC2 > .2) %>% 
  left_join(OTU.tax)
familynames <- stringr::str_split(active.spec.scores.pc$Family, "_")
fam.name.vec <- vector(length = length(active.spec.scores.pc$Family))
for(i in 1:length(active.spec.scores.pc$Family)){
  thisfamily <- familynames[[i]]
  fam.name.vec[i] <- paste(thisfamily[1], "sp.")
} 
active.spec.scores.pc$name <- fam.name.vec

scale.arrows0 <- 1.5

pc_labs <- full_join(design[which(design$sample.type == "RNA"),], active.traj.df) %>%
  filter(lubridate::month(date) %in% c(4, 7, 10, 1)) %>% 
  mutate(lab = paste(lubridate::month(date, label = T), lubridate::year(date))) %>% 
  group_by(lab) %>% 
  summarize(PC1mean = mean(PC1), PC2mean = mean(PC2))
  

fig.pca <- full_join(design[which(design$sample.type == "RNA"),], active.traj.df) %>% 
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_hline(aes(yintercept = 0), alpha = 0.1) + 
  geom_vline(aes(xintercept = 0), alpha = 0.1) +
  #stat_ellipse(aes(color = lubridate::month(date, label = T)), alpha = 0.5) +
  geom_path(alpha = 0.3, arrow = arrow(angle = 15, length = unit(0.1, "inches"),
                  ends = "last", type = "closed")) + 
  geom_point(aes(color = lubridate::month(date, label = T),
                 shape = lubridate::month(date, label = T)), alpha = 0.5, size = 3) +
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) +
  scale_shape_manual(values = rep(15:18,3)) +
  # geom_segment(data = active.spec.scores.pc,
  #              aes(x = 0, y = 0,
  #                  xend = scale.arrows0*PC1,
  #                  yend = scale.arrows0*PC2),
  #              alpha = 0.7,
  #              size = 1,
  #              color = "black",
  #              arrow = arrow(
  #                angle = 20,
  #                length = unit(.1, "inches"),
  #                type = "open"),
  #            show.legend = F) +
  # geom_text_repel(data = active.spec.scores.pc,
  #                  aes(x = scale.arrows0*PC1,
  #                      y = scale.arrows0*PC2),
  #                  label = active.spec.scores.pc$name,
  #                  # nudge_x = c(-.2, .4, .5, .4),
  #                  nudge_y = c(.1, 0.05, .05, 0),
  #                  point.padding = .5, alpha = 1) +
  xlab(paste("PC 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PC 2 (",var2,"%)", sep = "")) +
  coord_fixed() +
  geom_text_repel(data = pc_labs, mapping = aes(x = PC1mean, y = PC2mean, label = lab)) +
  ggsave("figures/fig1a.pdf", bg = "white", height = 6, width = 8) +
  ggsave("figures/fig1a.png", height = 6, width = 8, dpi = 600)
fig.pca

```

### Explaining temporal beta diversity

#### RDA
```{r}
#### Constrained Ordination
# samplingdays <- env.data$date
# intervals <- as.vector(diff.Date(samplingdays))
# w <- 1/(intervals/max(intervals))
# aem.out <- adespatial::aem.time(n = length(samplingdays), w = w, moran = TRUE)
# aem.eigs <- data.frame(aem.out$aem)
# aem.out$values
# colnames(aem.eigs) <- paste0("AEM",seq(1,ncol(aem.eigs)))
# 
env.preds <- na.omit(env.data[c("sample.id", "temp", "spc", "oxygen", "secchi", "ph", "tp", "tn", "doc")])
# 
OTUs.rel.act.subs <- OTUs.REL.active[design$sample.id[which(design$sample.type == "RNA")] %in% env.preds$sample.id,]
# aem.subs <- aem.eigs[design$sample.id[which(design$sample.type == "RNA")] %in% env.preds$sample.id,]
# 
# # #active.varpar <- varpart(OTUs.rel.act.subs, env.preds)
# active.rda.0 <- rda(OTUs.rel.act.subs ~ 1, data = aem.subs)
# # active.rda.1 <- rda(OTUs.rel.act.subs ~ ., data = aem.subs)
# # active.rda.2 <- rda(OTUs.rel.act.subs ~ AEM2 + AEM4 + AEM5 + AEM6, data = aem.subs)
# # active.rda <- ordiR2step(active.rda.0, scope = formula(active.rda.2), direction = "forward")
# # 
# # active.rda.env.0 <- rda(OTUs.rel.act.subs ~ 1, data = env.preds)
# # active.rda.env.1 <- rda(OTUs.rel.act.subs ~ ., data = env.preds)
# # active.rda.env <- ordiR2step(active.rda.env.0, scope = formula(active.rda.env.1), direction = "forward")
# 
# envfit(active.rda.0, aem.subs)
# 
# time.mod <- model.matrix(~ -1 + AEM1 + AEM2 + AEM3 + AEM4 + AEM5 + AEM6 + AEM7 + AEM15, data = aem.subs)

env.mod <- model.matrix(~ temp + doc + spc + secchi + tn + tp + ph, data = env.preds)

# RDA without controlling for time
active.rda <- rda(OTUs.rel.act.subs ~ env.mod)

active.rda.sum <- summary(active.rda)

env.vecs <- cbind.data.frame(center = 0, active.rda.sum$biplot[,c(1:2)])
rownames(env.vecs) <- c("Temp", "DOC", "Sp. Cond.", "Secchi depth", "TN", "TP", "pH")
spec.scores <- cbind.data.frame(center = 0, scores(active.rda)$species)
site.scores <- as.data.frame(scores(active.rda)$sites)


active.site.scores <- site.scores %>% rownames_to_column(var = "sample.name") %>% inner_join(design)
active.spec.scores <- spec.scores[which(abs(spec.scores$RDA1) > 0.2 | abs(spec.scores$RDA2) > 0.2),] %>% 
  rownames_to_column(var = "OTU") %>% inner_join(OTU.tax)

# active.spec.scores.rda <- as.data.frame(scores(active.rda, display = "species")) %>% rownames_to_column(var = "OTU") 
#   filter(RDA1 > .2 | RDA2 > .2) %>% 
#   left_join(OTU.tax)
familynames <- stringr::str_split(active.spec.scores$Family, "_")
fam.name.vec <- vector(length = length(active.spec.scores$Family))
for(i in 1:length(active.spec.scores$Family)){
  thisfamily <- familynames[[i]]
  fam.name.vec[i] <- paste(thisfamily[1], "sp.")
} 
active.spec.scores$name <- fam.name.vec
active.spec.scores$name <- fam.name.vec

var.explain1 <- round(eigenvals(active.rda)[1]/sum(eigenvals(active.rda))*100,2)
var.explain2 <- round(eigenvals(active.rda)[2]/sum(eigenvals(active.rda))*100,2)

scale.arrows <- 1.4
scale.arrows2 <- 1
fig.rda <- 
  ggplot() + 
  # Add site scores as points
  # geom_point(data = site.scores, aes(x = RDA1, y = RDA2)) + 
  geom_hline(aes(yintercept = 0), alpha = 0.1) + 
  geom_vline(aes(xintercept = 0), alpha = 0.1) +
  
  geom_point(data = active.site.scores, alpha = 0.5, size = 3,
            aes(x = RDA1, y = RDA2, shape = lubridate::month(date, label = T, abbr = T),
                color = lubridate::month(date, label = T, abbr = T))) + 
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.2), 9)) +
  scale_shape_manual(values = rep(15:18,3)) +
  # # Add species scores as vectors
  # geom_segment(data = active.spec.scores, 
  #              aes(x = center, y = center, 
  #                  xend = scale.arrows*RDA1, yend = scale.arrows*RDA2), 
  #              alpha = .5, color = "black",
  #            arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"), 
  #            show.legend = F) + 
  # geom_text_repel(data = active.spec.scores, 
  #              aes(x = (scale.arrows +.1)*RDA1, y = (scale.arrows+.1)*RDA2), 
  #              label = active.spec.scores$name, alpha = 1, point.padding = .25) + 

  # Add environmental vectors in blue
  geom_segment(data = env.vecs,
             aes(x = center, y = center, 
                 xend = scale.arrows2*RDA1, 
                 yend = scale.arrows2*RDA2),
             arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"),
             show.legend = F, alpha = .5, color = "black") +
  
  geom_text_repel(data = env.vecs, 
                  aes(x = (scale.arrows2)*RDA1, y = (scale.arrows2)*RDA2),
                    label = rownames(env.vecs), color = "black",
                   segment.alpha = 0.3, alpha = 1, point.padding = .25) +
  
   
  coord_fixed() + 
  scale_x_reverse() +
  labs(x = paste0("RDA 1 (",var.explain1,"%)"),
       y = paste0("RDA 2 (",var.explain2,"%)")) +
  ggsave("figures/fig1b.pdf", bg = "white", height = 6, width = 8) +
  ggsave("figures/fig1b.png", bg = "white", height = 6, width = 8, dpi = 600)
fig.rda

# RDA controlling for time
# active.rda <- rda(OTUs.rel.act.subs ~ env.mod + Condition(time.mod))
# 
# active.rda.sum <- summary(active.rda)
# 
# env.vecs <- cbind.data.frame(center = 0, active.rda.sum$biplot[,c(1:2)])
# rownames(env.vecs) <- c("Temp", "DOC", "DO", "SpC", "pH", "secchi")
# spec.scores <- cbind.data.frame(center = 0, scores(active.rda)$species)
# site.scores <- as.data.frame(scores(active.rda)$sites)
# 
# 
# active.site.scores <- site.scores %>% rownames_to_column(var = "sample.name") %>% inner_join(design)
# active.spec.scores <- spec.scores[which(abs(spec.scores$RDA1) > 0.06 | abs(spec.scores$RDA2) > 0.06),] %>% 
#   rownames_to_column(var = "OTU") %>% inner_join(OTU.tax)
# 
# # active.spec.scores.rda <- as.data.frame(scores(active.rda, display = "species")) %>% rownames_to_column(var = "OTU") 
# #   filter(RDA1 > .2 | RDA2 > .2) %>% 
# #   left_join(OTU.tax)
# familynames <- stringr::str_split(active.spec.scores$Family, "_")
# fam.name.vec <- vector(length = length(active.spec.scores$Family))
# for(i in 1:length(active.spec.scores$Family)){
#   thisfamily <- familynames[[i]]
#   fam.name.vec[i] <- paste(thisfamily[1], "sp.")
# } 
# active.spec.scores$name <- fam.name.vec
# active.spec.scores$name <- fam.name.vec
# 
# var.explain1 <- round(eigenvals(active.rda)[1]/sum(eigenvals(active.rda))*100,2)
# var.explain2 <- round(eigenvals(active.rda)[2]/sum(eigenvals(active.rda))*100,2)
# 
# scale.arrows <- 6
# scale.arrows2 <- 1.5
# fig.rda <- ggplot() + 
#   # Add site scores as points
#   # geom_point(data = site.scores, aes(x = RDA1, y = RDA2)) + 
#   geom_point(data = active.site.scores, alpha = 0.9, size = 3,
#             aes(x = RDA1, y = RDA2, 
#                 color = lubridate::month(date, label = T, abbr = T),
#                 shape = lubridate::month(date, label = T, abbr = T))) + 
#   scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) +
#   scale_shape_manual(values = c(0:2,5:14)) +
#   # Add species scores as vectors
#   geom_segment(data = active.spec.scores, 
#                aes(x = center, y = center, 
#                    xend = scale.arrows*RDA1, yend = scale.arrows*RDA2), 
#                alpha = .5, color = "black",
#              arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"), 
#              show.legend = F) + 
#   geom_text_repel(data = active.spec.scores, 
#                aes(x = scale.arrows*RDA1, y = scale.arrows*RDA2), 
#                label = active.spec.scores$name, alpha = 1, point.padding = .3) + 
# 
#   # Add environmental vectors in blue
#   geom_segment(data = env.vecs,
#              aes(x = center, y = center, 
#                  xend = scale.arrows2*RDA1, 
#                  yend = scale.arrows2*RDA2),
#              arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"),
#              show.legend = F, alpha = .5, color = "blue") +
#   geom_text_repel(data = env.vecs, 
#                   aes(x = scale.arrows2*RDA1, y = scale.arrows2*RDA2),
#                     label = rownames(env.vecs), color = "blue", 
#                    segment.alpha = 0.3, alpha = 1) + 
#   coord_fixed() + 
#   labs(x = paste0("RDA 1 (",var.explain1,"%)"),
#        y = paste0("RDA 2 (",var.explain2,"%)")) +
#   ggsave("figures/fig1b2.pdf", bg = "white", height = 6, width = 8)

# plot combined figure
plot_grid(fig.pca, fig.rda + theme(legend.position = "none"), align = "hv", axis = "lr", ncol = 1, labels = "AUTO") +
  ggsave("figures/fig1.pdf", height = 9.25, width = 5.5) +
  ggsave("figures/fig1.png", height = 9.25, width = 5.5, dpi = 700, bg = "white")

## so this plot shows, after controlling for temporal autocorrelation, which environmental drivers are most important for some taxa
```

# Is negative frequency dependence disproportionately stronger on rare than common taxa?
```{r}
# calc.neg.freq.dep <- function(sbs){
#   y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
#   for(i in 1:(nrow(sbs)-1)){
#     y.s[i,] = log(sbs[i+1,] / sbs[i,])
#   }
#   eq.freq <- vector(length = ncol(y.s))
#   fd <- vector(length = ncol(y.s))
#   sbs.rel <- decostand(sbs[-nrow(sbs),], method = "total")
#   for(i in 1:ncol(y.s)){
#     mod <- lm(y.s[,i] ~ sbs.rel[,i])
#     eq.freq[i] <- -coef(mod)[1] / coef(mod)[2]
#     fd[i] <- coef(mod)[2]
#   }
#   # plot(log10(eq.freq), log10(-fd), 
#   #      ylab = "log10 Negative Freq. Dependence",
#   #      xlab = "log10 Equilibrium Frequency")
#   return(na.omit(cbind.data.frame(eq.freq, fd)))
# }

# using mean rel abund
calc.neg.freq.dep <- function(sbs){
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  eq.freq <- vector(length = ncol(y.s))
  fd <- vector(length = ncol(y.s))
  sbs.rel <- decostand(sbs[-nrow(sbs),], method = "total")
  for(i in 1:ncol(y.s)){
    mod <- lm(y.s[,i] ~ sbs.rel[,i])
    eq.freq[i] <- mean(sbs.rel[,i])
    fd[i] <- coef(mod)[2]
  }
  # plot(log10(eq.freq), log10(-fd), 
  #      ylab = "log10 Negative Freq. Dependence",
  #      xlab = "log10 Equilibrium Frequency")
  return(na.omit(cbind.data.frame(eq.freq, fd)))
}


shuffle.ts <- function(species.ts){
  species.ts[sample(1:length(species.ts))]
}


get.null.distr <- function(comm, iters = 5000){
  pb <- progress_bar$new(total = iters)
  nfd.cov <- rep(NA, iters)
  for(i in 1:iters){
    nullcom <- apply(comm, MARGIN = 2, FUN = shuffle.ts) # randomize sampling order
    nfd <- calc.neg.freq.dep(nullcom)
    nfd <- nfd %>% filter(-fd > 0)
    nfd.cov[i] <- cov(log(nfd$eq.freq), log(-nfd$fd))
    pb$tick()
  }
  return(nfd.cov)
}

get.cov <- function(comm){
  nullcom <- apply(comm, MARGIN = 2, FUN = shuffle.ts) # randomize sampling order
  nfd <- calc.neg.freq.dep(nullcom)
  nfd <- nfd %>% filter(-fd > 0)
  return(cov(log(nfd$eq.freq), log(-nfd$fd)))
}

# OTUs.ranked <- OTUs.total[,names(sort(colSums(OTUs.total), decreasing = T))]



# How many taxa to consider?
ncol(OTUs.total.persistent)



OTUs.total.nozero <- replace(OTUs.total.persistent, which(OTUs.total.persistent == 0), .5)
nfd.total <- calc.neg.freq.dep(OTUs.total.nozero)
nfd.total %>% filter(-fd > 0) -> nfd.total
nfd.total.cov <- cov(log(nfd.total$eq.freq), log(-nfd.total$fd))
nfd.total %>% ggplot(aes(x = eq.freq, y = -fd)) +
  geom_point(alpha = 0.5) + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm", color = "black") + 
  ylab("Negative Frequency Dependence") + 
  xlab("Equilibrium Frequency") + NULL
  #ggsave("figures/NFD_total.pdf", width = 6, height = 3/4*6)

OTUs.active.nozero <- replace(OTUs.active.persistent, which(OTUs.active.persistent == 0), .5)
nfd.active <- calc.neg.freq.dep(OTUs.active.nozero)
nfd.active %>% filter(-fd > 0) -> nfd.active
nfd.active.cov <- cov(log(nfd.active$eq.freq), log(-nfd.active$fd))
nfdactivefig <- nfd.active %>% ggplot(aes(x = eq.freq, y = -fd)) +
  geom_point(alpha = 0.5, size = 3) + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm", color = "black") + 
  ylab("Strength of NFD") + 
  xlab("Equilibrium Frequency") + NULL
  #ggsave("figures/NFD_active.pdf", width = 6, height = 4.5)

rbind.data.frame(cbind.data.frame(nfd.active, subset = "Active"),
                 cbind.data.frame(nfd.total, subset = "Total")) %>% 
  ggplot(aes(x = eq.freq, y = -fd, color = subset, fill = subset)) +
  geom_point(alpha = 0.5) + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm") + 
  ylab("Negative Frequency Dependence") + 
  xlab("Equilibrium Frequency") + NULL
  #ggsave("figures/NFD_both.pdf", width = 6, height = 4.5)
 

# Raw NFD calculations
#raw.NFD <- function(sbs){
sbs = OTUs.active.nozero
y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
for(i in 1:(nrow(sbs)-1)){
  y.s[i,] = log(sbs[i+1,] / sbs[i,])
}
y.s <- as.data.frame(y.s)
colnames(y.s) <- colnames(sbs)
y.s$sample.id <- rownames(sbs)[-123]
sbs <- as.data.frame(decostand(sbs, method = "total"))
sbs <- rownames_to_column(sbs, var = "sample.id")
y.s <- gather(y.s, -sample.id, key = "OTU", value = "lambda")
sbs <- gather(sbs, -sample.id, key = "OTU", value = "abund")
joint_data <- left_join(y.s, sbs)

ggplot(joint_data, aes(x = abund, y = lambda)) + 
  geom_hline(yintercept = 0, alpha = 0.2) +
  geom_point(alpha = 0.3, size = 2) + 
  geom_smooth(method = "lm", size = 2) + 
  theme(legend.position = "none", axis.title = element_text(size = 32)) + 
  scale_x_log10() +
  facet_wrap(~OTU) +
  labs(x = "Relative abundance at time t",
       y = expression(Ln(N[t+1]/N[t])),
       title = "NFD for persistent OTUs when Active") + 
  ggsave("figures/figS3.pdf", height = 20*1.5, width = 30*1.5) +
  ggsave("figures/figS3.png", height = 20*1.5, width = 30*1.5, dpi = 500)

nfdactiveslopesfig <- ggplot(joint_data, aes(x = abund, y = lambda, color = OTU)) + 
  geom_hline(yintercept = 0, alpha = 0.2, linetype = "dashed") +
  #geom_point(alpha = 0.3) + 
  geom_line(stat = "smooth", method = "lm", se = F, alpha = 0.5, size = 1) + 
  theme(legend.position = "none") + 
  labs(x = "Relative abundance at time t",
       y = expression(Ln(N[t+1]/N[t]))) + 
  scale_color_discrete_qualitative(palette = "Harmonic") +
  scale_y_continuous(labels = c(0, -2, -4, -6), breaks = c(0, -2, -4, -6)) + NULL
  #ggsave("figures/NFD_OTU_slopes_active.pdf", height = 3/4*6, width = 6)

plot_grid(nfdactiveslopesfig, nfdactivefig, ncol = 1, labels = "AUTO") +
  ggsave("figures/figS4.pdf", width = 6, height = 8) +
  ggsave("figures/figS4.png", width = 6, height = 8, dpi = 600)

sbs = OTUs.total.nozero
y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
for(i in 1:(nrow(sbs)-1)){
  y.s[i,] = log(sbs[i+1,] / sbs[i,])
}
y.s <- as.data.frame(y.s)
colnames(y.s) <- colnames(sbs)
y.s$sample.id <- rownames(sbs)[-123]
sbs <- as.data.frame(decostand(sbs, method = "total"))
sbs <- rownames_to_column(sbs, var = "sample.id")
y.s <- gather(y.s, -sample.id, key = "OTU", value = "lambda")
sbs <- gather(sbs, -sample.id, key = "OTU", value = "abund")
joint_data <- left_join(y.s, sbs)
ggplot(joint_data, aes(x = abund, y = lambda)) + 
  geom_hline(yintercept = 0, alpha = 0.2) +
  geom_point(alpha = 0.3) + 
  geom_smooth(method = "lm") + 
  theme(legend.position = "none") + 
  facet_wrap(~OTU) +
  scale_x_log10() +
  labs(x = "Relative abundance at time t",
       y = expression(Ln(N[t+1]/N[t]))) + 
  NULL
  #ggsave("figures/NFD_individual_OTUs_total.pdf", height = 20*1.5, width = 30*1.5)

ggplot(joint_data, aes(x = abund, y = lambda, color = OTU)) + 
  geom_hline(yintercept = 0, alpha = 0.2, linetype = "dashed") +
  #geom_point(alpha = 0.3) + 
  geom_line(stat = "smooth", method = "lm", se = F, alpha = 0.5) + 
  theme(legend.position = "none") + 
  labs(x = "Relative abundance at time t",
       y = expression(Ln(N[t+1]/N[t]))) + 
  scale_color_discrete_qualitative(palette = "Harmonic") + 
  NULL
  #ggsave("figures/NFD_OTU_slopes_total.pdf", height = 3/4*6, width = 6)

```

```{r}
# Generate null distributions
iternum <- 5000

set.seed(82072)

# #setup parallel backend to use many processors
cores=detectCores(logical = F)
cl <- makeCluster(cores) #not to overload your computer
registerDoParallel(cl)

tot.nd <- foreach(i=1:iternum, .combine=rbind, .packages = c("vegan","tidyverse")) %dopar% {
   get.cov(OTUs.total.nozero)
}
act.nd <- foreach(i=1:iternum, .combine=rbind, .packages = c("vegan","tidyverse")) %dopar% {
   get.cov(OTUs.active.nozero)
}
#stop cluster
stopCluster(cl)

saveRDS(tot.nd, file = "analysis/temp-results/nfd_tot_null.rda")
saveRDS(act.nd, file = "analysis/temp-results/nfd_act_null.rda")


# Compare observed with null
hist(tot.nd, breaks = 20); abline(v = nfd.total.cov, col = "red")
(tot.nfd.pval <- rank(c(nfd.total.cov, tot.nd))[1] / length(c(nfd.total.cov, tot.nd)))
hist(act.nd, breaks = 20); abline(v = nfd.active.cov, col = "red")
(act.nfd.pval <- rank(c(nfd.active.cov, act.nd))[1] / length(c(nfd.active.cov, act.nd)))

nfd.obs <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = nfd.total.cov),
                 cbind.data.frame(subset = "Active", nfd = nfd.active.cov)) 

nfd.nulldists <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = tot.nd),
                 cbind.data.frame(subset = "Active", nfd = act.nd)) 

nfd.nulldists %>% 
  ggplot(aes(y = nfd, x = subset)) + 
  #geom_jitter(color = "grey", alpha = 0.1) + 
  geom_violin(alpha = 0.5, show.legend = F, fill = "grey") + 
  geom_point(data = nfd.obs, aes(x = subset, y = nfd), 
             color = "black", size = 5, show.legend = F) + 
  scale_fill_manual(values = my.cols[c(1,4)]) +
  coord_flip() +
  ylab("Covariance between Equilibrium Frequency\n and Negative Frequency Dependence") + 
  xlab("") + 
  ggsave("figures/fig3.pdf", bg = "white", width = 6, height = 4, units = "in") +
  ggsave("figures/fig3.png", bg = "white", width = 6, height = 4, dpi = 600)

# Calculate standardized effect sizes
nfd.summary <- nfd.nulldists %>% 
  group_by(subset) %>% summarize(mean.nfd = mean(nfd), sd.nfd = sd(nfd)) %>% 
  left_join(nfd.obs) %>% mutate(nfd.ses = (nfd - mean.nfd)/sd.nfd) %>% 
  mutate(nfd.strength = nfd / mean.nfd)
nfd.summary
```


# Reappeareance Scores
```{r}

reappearance.score <- function(DNA, RNA, quiet = TRUE){
  
  
  otu.dna.pa <- decostand(DNA, method = "pa")
  otu.rna.pa <- decostand(RNA, method = "pa")
  otu.diff <- otu.dna.pa - otu.rna.pa
    
  otuscores <- vector(length = ncol(otu.diff))
  
  # for each OTU in the difference OTU matrix
  for(j in 1:ncol(otu.diff)){
    
    # first assign a score of 0
    otuscore <- 0
    
    # then, go row by row within each OTU
    for(i in 1:(nrow(otu.diff)-1)){
      
      # each time an OTU reappears, score it 1
      if(otu.rna.pa[i, j] == 0 & # absent from RNA at i
         otu.dna.pa[i, j] == 1 & # present in DNA at i
         otu.rna.pa[i+1, j] == 1){ # present again in RNA
        otuscore <- otuscore + 1
      }
    }
    otuscores[j] <- otuscore
    if(quiet != TRUE) print(paste("completed otu #", j))
  }
  
  return(otuscores)
}

otu.reap <- reappearance.score(DNA = OTUs.total, RNA = OTUs.active, quiet = TRUE)
otu.pres <- colSums(OTUs.total.pa)

# Compute mean without zeroes
# nzmean <- function(x) {
#     if (all(x==0)) 0 else mean(x[x!=0])
# }
# 
# otu.total.relabund <- apply(decostand((OTUs.total), method = "total"),
#                             MARGIN = 2,
#                             FUN = nzmean)
# otu.active.relabund <- apply(decostand((OTUs.active), method = "total"),
#                             MARGIN = 2,
#                             FUN = nzmean)

overall_relabund <- function(comm){
  return( (colSums(comm)/sum(colSums(comm))) )
}

otu.total.relabund <- overall_relabund(OTUs.total)
otu.active.relabund <- overall_relabund(OTUs.active)


persistent_taxa <- colnames(OTUs.total.persistent)

persistent_seedbank_ornot <- cbind.data.frame(OTU = colnames(OTUs.total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
  filter(times_present > 0) %>% 
  filter(relabund_total > 0) %>% 
  filter(times_present > 1) %>% 
  mutate(percent_transition = times_activated/times_present) %>% 
  mutate(seedbank = ifelse(percent_transition == 0, F,T)) %>% 
  mutate(persistent = ifelse(OTU %in% persistent_taxa, T,F))

# taxa not relying on seed bank 
no_sb_taxa <- persistent_seedbank_ornot %>% filter(persistent == T) %>% 
  filter(seedbank == F) %>% 
  left_join(OTU.tax)
yes_sb_taxa <- persistent_seedbank_ornot %>% filter(persistent == T) %>% 
  filter(seedbank == T) %>% 
  left_join(OTU.tax)



reactivations <- cbind.data.frame(OTU = colnames(OTUs.total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
  filter(times_present > 0) %>% 
  filter(relabund_total > 0) %>% 
  filter(relabund_active > 0) %>% 
  filter(times_present > 1) %>% 
  mutate(percent_transition = times_activated/times_present) %>% 
  mutate(seedbank = ifelse(percent_transition == 0, "Not reliant on seed bank", "Persistent via seed bank")) %>% 
  mutate(persistent = ifelse(OTU %in% persistent_taxa, "Persistent", "Non-persistent")) %>% 
  filter(persistent == "Persistent") %>% 
  gather(relabund_active, relabund_total, key = "subset", value = "relabund") %>% 
  mutate(subset = ifelse(subset == "relabund_active", "Active Portion", "Total Community")) %>% 
  filter(subset == "Active Portion") 


exp.mod <- nls(times_activated ~ b*exp(-alpha*log10(relabund)), data = reactivations, start = list(alpha=1, b = 1))
plot(exp.mod)
summary(exp.mod)

reactivations %>% 
  ggplot(aes(y = times_activated, x = relabund)) + 
  geom_point(size = 3, alpha = 0.5) + 
  geom_line(data = augment(exp.mod), aes(x = relabund, y = .fitted), size = 1) + 
  #geom_smooth(color = "black", span = 1) + 
  #facet_wrap(~subset, ncol = 1) +
  theme(legend.position = "bottom", legend.text = element_text(size = 10)) +
  ylab("Reactivations from seed bank") +
  xlab("Relative abundance") +
  scale_x_log10() +
  ggsave("figures/fig5.pdf", width = 6, height = 3/4*6) +
  ggsave("figures/fig5.png", width = 6, height = 3/4*6, dpi = 600)


obs_dat <- cbind.data.frame(OTU = colnames(OTUs.total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
  filter(times_present > 0) %>% 
  filter(relabund_total > 0) %>% 
  filter(relabund_active > 0) %>% 
  filter(times_present > 1) %>% 
  mutate(percent_transition = times_activated/times_present) %>% 
  mutate(seedbank = ifelse(percent_transition == 0, "Not reliant on seed bank", "Persistent via seed bank")) %>% 
  mutate(persistent = ifelse(OTU %in% persistent_taxa, "Persistent", "Non-persistent")) %>% 
  filter(persistent == "Persistent") %>% 
  gather(relabund_active, relabund_total, key = "subset", value = "relabund") %>% 
  mutate(subset = ifelse(subset == "relabund_active", "Active Portion", "Total Community")) %>% 
  filter(subset == "Active Portion")
  
# pull out taxa who are present > 50% of time but inactive > 50% of time
# these may be potential seed bankers or sink populations form soils
cbind.data.frame(times_inactive = otu.reap, presences = otu.pres) %>%
  rownames_to_column(var = "OTU") %>% 
  mutate(prop_inactive = times_inactive/presences, 
         incidence = presences / (length(env.data$date))) %>% 
  filter(incidence > 0.9, prop_inactive > 0.1) -> seedbank.taxa
subset(OTU.tax, OTU %in% seedbank.taxa$OTU)
```

# compare with a null model
```{r}
reap.nullmod <- function(DNAcomm = OTUs.total.persistent,
                         RNAcomm = OTUs.active.persistent,
                         samp.design = design, n = 10, algorithm = "c0_samp"){
  
  null_correlations <- numeric(length = n)
  null_data <- data.frame()
  n_taxa <- ncol(DNAcomm)
  n_site <- nrow(DNAcomm)
  
  # define null model
  DNAnullcom <- nullmodel(DNAcomm, method = algorithm)
  #RNAnullcom <- nullmodel(RNAcomm, method = algorithm)
  
  # sim
  print("Generating DNA null models.")
  DNAsims <- as.array(simulate(DNAnullcom, nsim = n, seed = 1))
  
  print("Generating RNA null models.")
  #RNAsims <- as.array(simulate(RNAnullcom, nsim = n, seed = 1))
  RNAsims <- DNAsims * 0
  
  # loop through n sims
  for(i in 1:n){
    
    # loop across taxa
    for(j in 1:n_taxa){
      n_active <- sum(RNAcomm[,j] > 0)
      n_present <- sum(DNAcomm[,j]>0)
      
      # create vector of active and present, then randomly sample
      v <- sample(c(rep(1, n_active), rep(0,(n_present-n_active))))
      
      # then add the elements only to the positions where DNA is present
      RNAsims[which(DNAsims[,j,i] > 0),j,i] <- v
      
    }  
  }
  
  
  # loop through null models
  print("Analyzing sim output:")
  for(i in 1:n){
    
    nullcom_active <- RNAsims[,,i]
    nullcom_total <- DNAsims[,,i]
    
    
    otu.reap <- reappearance.score(DNA = nullcom_total, RNA = nullcom_active)
    otu.pres <- colSums(nullcom_total > 0)
    
    # otu.total.relabund <- apply(decostand((nullcom_total), method = "total"),
    #                         MARGIN = 2,
    #                         FUN = nzmean)
    # otu.active.relabund <- apply(decostand((nullcom_active), method = "total"),
    #                         MARGIN = 2,
    #                         FUN = nzmean)
    
    otu.total.relabund <- overall_relabund(nullcom_total)
    otu.active.relabund <- overall_relabund(nullcom_active)


    to_correlate <- cbind.data.frame(OTU = colnames(nullcom_total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
    filter(times_present > 0) %>% 
    filter(relabund_total > 0) %>% 
    filter(relabund_active > 0) %>% 
    filter(times_present > 1) %>% 
    mutate(percent_transition = times_activated/times_present) %>% 
    mutate(seedbank = ifelse(percent_transition == 0, "Not reliant on seed bank", "Persistent via seed bank")) %>% 
    mutate(persistent = ifelse(OTU %in% persistent_taxa, "Persistent", "Non-persistent")) %>% 
    filter(persistent == "Persistent") %>% 
    gather(relabund_active, relabund_total, key = "subset", value = "relabund") %>% 
    mutate(subset = ifelse(subset == "relabund_active", "Active Portion", "Total Community")) %>% 
    filter(subset == "Active Portion")  
    
    
    # to_correlate %>%
    # ggplot(aes(y = times_activated, x = relabund)) +
    # geom_jitter(size = 3, alpha = 0.5) +
    # geom_smooth(color = "black", method = "lm") +
    # #facet_wrap(~subset, ncol = 1) +
    # theme(legend.position = "bottom", legend.text = element_text(size = 10)) +
    # ylab("Reactivations from seed bank") +
    # xlab("Average relative abundance") +
    # scale_x_log10()
    print(paste("Completed sim", i, "of", n,"."))
    
    null_data <- rbind.data.frame(null_data, cbind.data.frame(sim = i, to_correlate))
    null_correlations[i] <- cor(to_correlate$times_activated, to_correlate$relabund)
    gc()

  }
  
  return(list(corr = null_correlations,
              dat = null_data))
}

# # Run Sims
# null_dist_tot <- list(corr = vector(), dat = data.frame())
# 
# nsims = 1000
# while(nsims > 0){
#   # run this batch of sims
#   sims_to_run <- min(100, nsims)
#   null_dist_i <- reap.nullmod(n = sims_to_run, algorithm = "c0_samp")
# 
#   null_dist_tot$corr <- c(null_dist_tot$corr, null_dist_i$corr)
#   null_dist_tot$dat <- rbind.data.frame(null_dist_tot$dat, null_dist_i$dat)
# 
#   nsims <- nsims - sims_to_run
#   print(paste("There are", nsims, "simulations remaining."))
# }
# 
# saveRDS(null_dist_tot, "analysis/null_dists_reactivations_twostep.rda")
null_dist_tot <- readRDS("analysis/null_dists_reactivations_twostep.rda")


obs_relabunds <- data.frame(relabund = overall_relabund(OTUs.active.persistent)) %>% 
  rownames_to_column(var = "OTU")

null_dist_toplot <- null_dist_tot$dat %>% 
  mutate(sim = as.character(sim)) %>% 
  bind_rows(bind_cols(sim = "observed", obs_dat)) %>% 
  mutate(obsim = ifelse(sim == "observed", "Observed", "Simulated")) 

null_calc_meansd <- null_dist_toplot %>% 
  filter(obsim == "Simulated") %>% 
  group_by(OTU) %>% 
  mutate(taxon_mean = mean(times_activated),
         taxon_sd = sd(times_activated)) %>% 
  select(OTU, taxon_mean, taxon_sd) %>% 
  unique(.)

null_dist_toplot_ses <- null_dist_toplot %>% 
  filter(obsim == "Observed") %>% 
  mutate(observed_reactivations = times_activated) %>% 
  select(OTU, observed_reactivations) %>% 
  left_join(obs_relabunds, by = "OTU") %>% 
  right_join(null_calc_meansd, by = "OTU") %>% 
  mutate(taxon_ses = (observed_reactivations - taxon_mean)/taxon_sd)
null_dist_toplot_ses %>% 
  ggplot(aes(x = relabund, y = taxon_ses, label = OTU)) + 
  geom_hline(yintercept = 2, alpha = 0.25, linetype = "dashed") +
  geom_hline(yintercept = -2, alpha = 0.25, linetype = "dashed") + 
  geom_point(alpha = .8, aes(color = abs(taxon_ses) > 2)) +
  geom_text_repel(data = . %>% filter(abs(taxon_ses) > 2)) +
  scale_x_log10() +
  theme(legend.position = "none") + 
  labs(x = "Overall relative abundance",
       y = "Standardized Effect Size \nof Reactivation Number") +
  ggsave("figures/figS5.pdf", width = 8, height = 3/4*8) +
  ggsave("figures/figS5.png", width = 8, height = 3/4*8, dpi = 600)

merge(reactivations, select(null_dist_toplot_ses, OTU, taxon_ses)) %>% 
  mutate(sig = ifelse((taxon_ses) < -2, "Significant (|SES| > 2)", "N.S. (|SES| < 2)")) %>% 
  mutate(sig = ifelse(is.na(sig), "Indeterminate", sig)) %>% 
  mutate(sig = factor(sig, levels = c("N.S. (|SES| < 2)", "Significant (|SES| > 2)", "Indeterminate"))) %>% 
  ggplot(aes(y = times_activated, x = relabund)) + 
  geom_point(aes(color = sig), size = 3, alpha = 0.8) + 
  geom_line(data = augment(exp.mod), aes(x = relabund, y = .fitted), size = 1) + 
  #geom_smooth(color = "black", span = 1) + 
  #facet_wrap(~subset, ncol = 1) +
  theme(legend.position = c(.86,.89), legend.text = element_text(size = 10)) +
  ylab("Reactivations from seed bank") +
  xlab("Average relative abundance") +
  #geom_text_repel(data = . %>% filter(abs(taxon_ses) > 2), aes(label = OTU), max.overlaps = 15) +
  scale_x_log10() +
  scale_color_discrete_qualitative(palette = "Harmonic") +
  ggsave("figures/fig5annotated.pdf", width = 8, height = 3/4*8) +
  ggsave("figures/fig5annotated.png", width = 8, height = 3/4*8, dpi = 600)


```


```{r}
maxgrowth <- function(sbs){
  
  # create new matrix to calculate growth rates
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  
  # for the gr's just calculated (ys) where is the max
  # for each OTU (i.e., each col) in the matrix
  max.growth.time <- apply(y.s, 2, which.max)
  
  # now, what is the gr at that max id'ed above
  max.growths <- apply(y.s, 2, max)
  
  # here, export a df of max gr and time of max gr for each OTU in the input df
  return(data.frame(
    rate = max.growths, sample = max.growth.time))
}

env.vars <- env.data[c("temp", "spc", "secchi", "ph", "tp", "tn", "doc")]
env.dist <- cluster::daisy(scale(env.vars), metric = "gower")
env.pca <- prcomp(na.omit(env.vars), scale. = T)

env.pcs <- as.data.frame(env.pca$x) %>% rownames_to_column(var = "sample.id") 
env.pcs$sample.id <- as.numeric(env.pcs$sample.id)
env.pcs <- env.pcs %>% right_join(env.data)

active.max.growths <- maxgrowth(OTUs.active.nozero)
active.max.growths$OTU <- colnames(OTUs.active.nozero)
active.max.growths <- active.max.growths %>% mutate(sample.id = sample)

env.pcs.subset <- subset(env.pcs, sample.id %in% active.max.growths$sample.id) %>% inner_join(active.max.growths) %>%
  left_join(OTU.tax)

env.pcs.subset %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_jitter(aes(
    color = lubridate::month(date, label = T, abbr = T)), 
    alpha = 0.5, width = .15, height = .15, size = 2) +
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) +
  geom_text_repel(data = subset(env.pcs.subset, 
                                (PC1 > 4 | PC1 < -1) | (abs(PC2) > 2)), 
                  aes(label = OTU, 
                      color = lubridate::month(date, label = T, abbr = T)), show.legend = F) +
  geom_segment(data = as.data.frame(env.pca$rotation),
               aes(x = 0, y = 0, xend = 3*PC1, yend = 3*PC2),
               arrow = arrow(length = unit(.3, "cm")),
               show.legend = F, alpha = 1, color = "blue") +
  geom_text_repel(data = as.data.frame(env.pca$rotation), 
            aes(x = 3*PC1, y = 3*PC2, label = rownames(env.pca$rotation)),
            box.padding = .5, point.padding = .5, alpha = 0.9,
            segment.alpha = 0) +
  coord_fixed()

getgrs <-  function(sbs){
  # create new matrix to calculate growth rates
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  return(y.s)
}
persistent.grs <- getgrs(OTUs.active.nozero)
colnames(persistent.grs) <- colnames(OTUs.active.persistent)
rownames(persistent.grs) <- rownames(OTUs.active.persistent)[-1]


persistent.grs.all <- as.data.frame(persistent.grs) %>% 
  rownames_to_column(var = "sample.name") %>% 
  gather(-"sample.name", key = "OTU", value = "growth") %>% 
  left_join(design) %>% 
  mutate(week = lubridate::week(date)) %>% 
  group_by(OTU, week) %>% 
  summarize(mean.growth = mean(growth),
            sd.growth = sd(growth))
maxgrs <- persistent.grs.all %>% 
  summarize(max.growth = max(mean.growth))

persistent.grs.all <- persistent.grs.all %>% 
  left_join(maxgrs) %>%
  left_join(OTU.tax)
# persistent.grs.all <- as.data.frame(persistent.grs) %>% 
#   rownames_to_column(var = "sample.name") %>% 
#   gather(-"sample.name", key = "OTU", value = "growth") %>% 
#   left_join(design) %>% 
#   left_join(OTU.tax)

persistent.grs.all <- persistent.grs.all %>% 
  mutate(ismax = ifelse(mean.growth == max.growth, T, F)) 

# this normalizes to jan 1 of each year for the avg growth rates.
persistent.grs.all$date <- lubridate::parse_date_time(paste(2014, persistent.grs.all$week, 'Wed', sep="/"),'Y/W/a') %>% as_date(.) %>% - 7
persistent.grs.all$Genus <- str_replace(persistent.grs.all$Genus, coll("_unclassified"), " sp.")
peaks_in_df <- persistent.grs.all %>% 
  filter(ismax) %>% 
  mutate(peaks_in = as_date(date)) %>% 
  select(OTU, peaks_in)
persistent.grs.all <- persistent.grs.all %>% 
  mutate(mean.growth = mean.growth / 7,
         max.growth = max.growth / 7) # convert to /day
persistent.grs.all <- persistent.grs.all %>%  
  left_join(peaks_in_df) 
seasonal_cols <- shifter(darken(plasma(12, direction = -1),.1))
dup_end <- persistent.grs.all %>% filter(date == min(date)) %>% 
  mutate(date = date + 364)
persistent.grs.all %>% 
  bind_rows(dup_end) %>% 
  #mutate(mean.growth = ifelse(mean.growth<0, 0, mean.growth)) %>% 
  ggplot(aes(x = date, group = OTU)) +
  geom_line(aes(y = mean.growth), alpha = 0.1, color = "gray70") + 
  geom_line(data = . %>% filter(OTU == "Otu00018"),
            aes(y = mean.growth, color = lubridate::month(peaks_in, label = T))) +
  geom_line(data = . %>% filter(OTU == "Otu00017"),
            aes(y = mean.growth, color = lubridate::month(peaks_in, label = T))) +
  # geom_line(data = . %>% filter(OTU == "Otu00022"),
  #           aes(y = mean.growth, color = lubridate::month(peaks_in, label = T))) +
  geom_line(data = . %>% filter(OTU == "Otu00001"),
            aes(y = mean.growth, color = lubridate::month(peaks_in, label = T))) +
  geom_point(data = subset(persistent.grs.all, ismax == T), 
             aes(y = max.growth, 
                 color = lubridate::month(date, label = T)), alpha = 0.7, size = 3) + 
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9), 
                     breaks = c("Jan", "Feb", "Mar", "Apr", "May", "Jun",
                                "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")) +
  #coord_flip() + 
  # geom_ribbon(aes(group = NULL, col = NULL, ymax = 0, ymin = -5), 
  #             fill = "white", col = "white", alpha = 0.8) +
  theme(legend.position = "bottom", legend.text = element_text(size = 9)) +
  # geom_text_repel(data = subset(persistent.grs.all, ismax == T),
  #           aes(y = max.growth, label = Genus), 
  #           nudge_y = 0, box.padding = .5, max.iter = 10000,
  #           size = 2, color = "black", segment.alpha = .5) +
  labs(x = "", 
       y = expression(Mean~growth~rate~(d^-1))) +
  scale_x_date(date_labels = "%b", breaks = "1 month", 
               labels = c("Jan", "Feb", "Mar", "Apr", "May", "Jun",
                           "Jul", "Aug", "Sep", "Oct", "Nov", "Dec")) +
  coord_polar(theta = "x") +
  theme_bw() +
  theme(legend.position = "none", 
        panel.grid.minor.y = element_blank(),
        #panel.grid.major.y = element_blank(),
        panel.grid.minor.x = element_blank()) +
  #scale_y_continuous(limits = c(-5, 12)) +
  ggsave("figures/fig2.pdf", bg = "white", width = 4.5, height = 4.5) +
  ggsave("figures/fig2.png", bg = "white", width = 4.5, height = 4.5, dpi = 600)
  # ggsave("figures/max_growth_date.png", bg = "white", width = 12, height = 16, dpi = 600)

persistent.grs.all %>%
  filter(ismax == T) %>% 
  ggplot(aes(x = date, y = max.growth, group = OTU)) +
  geom_point() +
  labs(x = "Week of the year", 
       y = expression(Mean~growth~rate~(d^-1))) +
  scale_color_discrete_qualitative(palette = "Dynamic") +
  scale_x_date(date_labels = "%b") +
  coord_polar(theta = "x") +
  theme_minimal()

active.max.growths %>% 
  mutate(rate = rate/7)  %>%  # convert to per day
  left_join(subset(design, sample.type == "RNA")) %>% 
  mutate(month = lubridate::month(date, label = T, abbr = F)) %>% 
  mutate(rate = round(rate, 3)) %>% 
  left_join(OTU.tax) %>% 
  arrange(month) %>% 
  write_csv("persistent_taxa.csv", col_names = T)
  
```

```{r}
library(ape)
library(ggtree)
library(aplot)

tree <- read.tree("data/ul_resgrad.tree")
summary(tree)

# drop all but persistetn OTUs
persistent.tree <- drop.tip(tree, 
         tree$tip.label[!(tree$tip.label %in% persistent.grs.all$OTU)])
persistent.max.grs <- subset(persistent.grs.all, ismax == T) %>% 
  rename(label = OTU)

tree_annotated <- ggtree(persistent.tree) %<+% persistent.max.grs

# tree
tree_plot <- tree_annotated + 
  theme_tree() +
  geom_tiplab(aes(color = lubridate::month(date, label = T)),
              offset = .05, align = T, show.legend = F) +
  geom_tippoint(aes(color = lubridate::month(date, label = T)), size = 2) +
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) + 
  theme(legend.title = element_blank(),
        legend.position = "none") +
  xlim_tree(xlim = 1.68) +
  ggsave("figures/tree_persistent_taxa.pdf", height = 12, width = 12)

# mean growth per month for the taxa
growth_plot <- persistent.grs.all %>% 
  mutate(month = lubridate::month(date, label = T)) %>%
  rename(label = OTU) %>% 
  ggplot(aes(x = month, y = label, fill = mean.growth)) + 
  geom_tile() + 
  scale_fill_continuous_diverging(palette = "Purple-Green", rev = F) +
  labs(x = "", y = "", fill = "Mean change in \nrel. abund. (per week)") +
  theme_tree2()

composite_plot <- growth_plot %>% insert_left(tree_plot)
ggsave(plot = composite_plot, filename = "figures/tree_growth_composite.pdf", width = 12, height = 12)


```

```{r}
env.pcs

cbind.data.frame(env.vars, decostand(OTUs.active.persistent, method = "total")) %>% 
  gather(-temp, -spc, -secchi, -ph, -tp, -tn, -doc, key = "OTU", value = "Abundance") %>% 
  gather(temp, spc, secchi, ph, tp, tn, doc, key = "Variable", value = "Value") %>% 

  ggplot(aes(x = Value, y = log10(Abundance), color = OTU)) + 
  geom_point() + 
  geom_smooth(se=F, span = 0.05) +
  facet_wrap(~Variable, scales = "free")

set.seed(123)
dat_for_mod <- cbind.data.frame(env.vars['temp'], decostand(OTUs.active.persistent, method = "total")[,sample(1:ncol(OTUs.active.persistent),25)])


set.seed(123)
cbind.data.frame(env.vars['temp'], decostand(OTUs.active.persistent, method = "total")[,sample(1:ncol(OTUs.active.persistent),25)]) %>% 
  gather(-temp, key = "OTU", value = "Abundance") %>% 
  gather(temp, key = "Variable", value = "Value") %>% 

  ggplot(aes(x = Value, y = log10(Abundance), color = OTU)) + 
  geom_point(alpha = 0.5) + 
  geom_line(stat = "smooth", method = "lm", formula = y ~ x + I(x^2), se=F, span = 0.5, alpha = 0.8) +
  facet_wrap(~OTU, scales = "free") +
  labs(x = "Temperature", y = expression(paste(Log[10], "(Relative Abundance)")))  +
  theme(legend.position = "none")+
  ggsave("figures/figS2.pdf", width = 5*4, height = 5*3) +
  ggsave("figures/figS2.png", width = 5*4, height = 5*3, dpi = 600)

set.seed(123)
cbind.data.frame(env.vars['temp'], decostand(OTUs.total.persistent, method = "total")[,sample(1:ncol(OTUs.active.persistent),25)]) %>% 
  gather(-temp, key = "OTU", value = "Abundance") %>% 
  gather(temp, key = "Variable", value = "Value") %>% 

  ggplot(aes(x = Value, y = log10(Abundance), color = OTU)) + 
  geom_point(alpha = 0.5) + 
  geom_line(stat = "smooth", method = "lm", formula = y ~ x + I(x^2), se=F, span = 0.5, alpha = 0.8) +
  facet_wrap(~OTU, scales = "free") +
  labs(x = "Temperature", y = expression(paste(Log[10], "(Relative Abundance)")))  +
  theme(legend.position = "none")#+
  #ggsave("figures/figS4_taxa-by-env.pdf", width = 5*4, height = 5*3)

```

