---
title: "University Lake Microbial Seed Bank"
author: "Nathan Wisnoski, Jay Lennon"
date: "1/19/2018"
output: html_document
---

```{r setup, include=FALSE}
# Load packages

require(vegan)
require(tidyverse)
require(lubridate)
require(viridis)
require(betapart)
require(cowplot)
#require(xts)
require(progress)
#require(cluster)
require(adespatial)
require(ggrepel)
require(colorspace)

source("bin/mothur_tools.R")

cv <- function(x){
  sd(x) / abs(mean(x))
}

shifter <- function(x, n = 1) {
  if (n == 0) x else c(tail(x, -n), head(x, n))
}
```

Next, we'll set the aesthetics of the figures we will produce. 
```{r figure_setup}
#my.cols <- RColorBrewer::brewer.pal(n = 4, name = "Greys")[3:4]

# Set theme for figures in the paper
theme_set(theme_classic() + 
  theme(axis.title = element_text(size = 16),
        axis.title.x = element_text(margin = margin(t = 15, b = 15)),
        axis.title.y = element_text(margin = margin(l = 15, r = 15)),
        axis.text = element_text(size = 14),
        axis.text.x = element_text(margin = margin(t = 5)),
        axis.text.y = element_text(margin = margin(r = 5)),
        #axis.line.x = element_line(size = 1),
        #axis.line.y = element_line(size = 1),
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.x = element_line(size = 1),
        axis.ticks.y = element_line(size = 1),
        axis.ticks.length = unit(.1, "in"),
        panel.border = element_rect(color = "black", fill = NA, size = 1.5),
        legend.title = element_blank(),
        legend.text = element_text(size = 14),
        strip.text = element_text(size = 14),
        strip.background = element_blank()
        ))
```

## Import Data
Here, we read in the processed sequence files from mothur (shared and taxonomy) and a design of the sampling. We also load in the environmental data. We then remove the mock community from the dataset and ensure the the design and OTU table are aligned by row.
```{r}
# Define Inputs
# Design = general design file for experiment
# shared = OTU table from mothur with sequence similarity clustering
# Taxonomy = Taxonomic information for each OTU
design <- "data/UL_timeseries_design.csv"
shared <- "data/ul_resgrad.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.pick.opti_mcc.shared"
taxon  <- "data/ul_resgrad.trim.contigs.good.unique.good.filter.unique.precluster.pick.pick.pick.opti_mcc.0.03.cons.taxonomy"

# Import Design
design <- read_csv(file = design)

# Import Shared Files
OTUs <- read.otu(shared = shared, cutoff = "0.03")    # 97% Similarity

# Import Taxonomy
OTU.tax <- read.tax(taxonomy = taxon, format = "rdp")


# Load environmental data
env.data <- read.table("data/ul-seedbank.env.txt", sep="\t", header=TRUE)
env.data$date <- as.Date(parse_date_time(env.data$date, "m d y"))
env.data$doc[which(env.data$doc == "**")] <- NA
env.data$doc <- as.numeric(env.data$doc)
summary(env.data)
design <- left_join(design, env.data[,c("sample.id", "date")])
env.data <- env.data[-which(!(env.data$date %in% design$date)),]

# Subset to just the UL timeseries samples
OTUs <- OTUs[str_which(rownames(OTUs), "UL"),]

# make sure OTU table matches up with design order
OTUs <- OTUs[match(design$sample.name, rownames(OTUs)),]
```

# transform OTUs
```{r data format, include=FALSE}

# Make rel abund matrices and split into active total comms
# OTUs <- OTUs[,-which(colSums(OTUs) < 5)]
set.seed(47405)
OTUs <- rrarefy(OTUs, sample = min(rowSums(OTUs)))

OTUs.active <- OTUs[which(design$sample.type == "RNA"),]
OTUs.total <- OTUs[which(design$sample.type == "DNA"),]
OTUs.total <- OTUs.total + decostand(OTUs.active, method = "pa")
OTUs.total.pa <- decostand(OTUs.total, method = "pa")
OTUs.active.pa <- decostand(OTUs.active, method = "pa")

#OTUs.REL <- decostand(OTUs, method = "total")
OTUs.REL <- decostand(OTUs, method = "hellinger")
OTUs.REL.total <- OTUs.REL[which(design$sample.type == "DNA"),]
OTUs.REL.active <- OTUs.REL[which(design$sample.type == "RNA"),]


# Define color pallette
my.cols <- viridis(4, begin = .3, end = 0.8)
```

# Characterize temporal dynamics of the bacterioplankton community 

## How does alpha diversity vary over time?
```{r}
alpha.div.total <- rowSums(OTUs.total.pa) #exp(diversity(OTUs.total, index = "shannon")) # spec equivs
alpha.div.active <- rowSums(OTUs.active.pa) #exp(diversity(OTUs.active, index = "shannon"))
data.frame(date = env.data$date, Total = alpha.div.total, Active = alpha.div.active) %>% 
  mutate(diversity_ratio = Total / Active) %>% 
  #gather(community, diversity, -date) %>% 
  ggplot(aes(x = date, y = diversity_ratio)) + 
  geom_vline(xintercept = as.numeric(as.Date(c("2014-01-01", "2015-01-01"))),
             linetype = 2, color = "grey", alpha = 0.8) +
  geom_line(size = 1, alpha = .8, show.legend = F) +
  geom_point(size = 3, alpha = 0.5) +
  geom_smooth(alpha = 0.3, color = "black", span = .5) +
  ylab(expression(paste(alpha,"-diversity ratio (total / active richness)"))) +
  xlab("") +
  scale_x_date(date_breaks = "3 month", 
               date_minor_breaks = "1 month", 
               date_labels = "%b %Y") +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  #scale_color_manual("Community Subset", values = c("#b2df8a", "#1f78b4")) +
  scale_color_manual(values = my.cols[c(4,1)]) +
  #scale_y_log10() +
  #annotation_logticks(sides = "l") +
  ggsave("figures/alpha_diversity.pdf", bg = "white", height = 6, width = 8) +
  ggsave("figures/alpha_diversity.png", bg = "white", height = 6, width = 8, dpi = 600)
```


## How does community composition change over time?
### Ordination methods (abundance-based)
```{r}
# # never active
# transients <- OTUs.REL.total[,which(colSums(OTUs.REL.active) == 0 & colSums(OTUs.REL.total) != 0)]
# transients.dist <- vegdist(transients, method = "euclidean")
# transients.pcoa <- cmdscale(transients.dist, eig = T)
# var1 <- round(eigenvals(transients.pcoa)[1] / sum(eigenvals(transients.pcoa)), 3) * 100
# var2 <- round(eigenvals(transients.pcoa)[2] / sum(eigenvals(transients.pcoa)), 3) * 100
# transients.traj <- transients.pcoa$points
# colnames(transients.traj) <- c("Comp.1", "Comp.2")
# transients.traj.df <- cbind.data.frame(
#   sample.id = design$sample.id[which(design$sample.type == "RNA")], transients.traj) 
# 
# 
# filter(design, sample.type == "RNA") %>% 
#   full_join(transients.traj.df) %>% 
#   mutate(month = lubridate::month(date, label = T, abbr = T)) %>% 
#   ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
#   geom_point() + 
#   scale_color_viridis(discrete = T, "Month") + 
#   theme_cowplot() + 
#   xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
#   ylab(paste("PCoA 2 (",var2,"%)", sep = "")) + 
#   coord_fixed() +
#   stat_ellipse()
# 
# 
# # present, but not active
# seedbank <- OTUs.REL.active == 0 & OTUs.REL.total != 0
# seedbank <- seedbank * OTUs.REL.total
# seedbank <- seedbank[,which(colSums(seedbank) > 0)]
# # remove transients
# seedbank <- seedbank[, -which(colnames(seedbank) %in% colnames(transients))]
# 
# seedbank.dist <- vegdist(seedbank, method = "euclidean")
# seedbank.pcoa <- cmdscale(seedbank.dist, eig = T)
# var1 <- round(eigenvals(seedbank.pcoa)[1] / sum(eigenvals(seedbank.pcoa)), 3) * 100
# var2 <- round(eigenvals(seedbank.pcoa)[2] / sum(eigenvals(seedbank.pcoa)), 3) * 100
# seedbank.traj <- seedbank.pcoa$points
# colnames(seedbank.traj) <- c("Comp.1", "Comp.2")
# 
# seedbank.traj.df <- cbind.data.frame(
#   sample.id = design$sample.id[which(design$sample.type == "RNA")], seedbank.traj) 
# filter(design, sample.type == "RNA") %>% 
#   full_join(seedbank.traj.df) %>% 
#   mutate(month = lubridate::month(date, label = T, abbr = T)) %>%
#   ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
#   geom_point(alpha = 0.5) +
#   # geom_path(arrow = arrow(angle = 15, length = unit(0.1, "inches"),
#   #                         ends = "last", type = "closed")) + 
#   scale_color_viridis(discrete = T, "Month") + 
#   geom_point(aes(x = seedbank.traj.df$Comp.1[1], 
#                  y = seedbank.traj.df$Comp.2[1]),
#              color = viridis(2)[1], size = 3) +
#   geom_point(aes(x = seedbank.traj.df$Comp.1[123], 
#                  y = seedbank.traj.df$Comp.2[123]),
#              color = viridis(2)[2], size = 3) +
#   theme_cowplot() + 
#   xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
#   ylab(paste("PCoA 2 (",var2,"%)", sep = "")) +
#   coord_fixed() +
#   stat_ellipse() +
#   ggsave("figures/seedbank_trajectory.pdf", bg = "white", height = 8, width = 10)
# 
# filter(design, sample.type == "RNA") %>% 
#   full_join(seedbank.traj.df) %>% 
#   mutate(month = lubridate::month(date, label = T, abbr = T)) %>% 
#   ggplot(aes(x = Comp.1, y = Comp.2, color = month)) + 
#   geom_point() + 
#   scale_color_viridis(discrete = T, "Month") + 
#   theme_cowplot() + 
#   xlab(paste("PCoA 1 (",var1,"%)", sep = "")) + 
#   ylab(paste("PCoA 2 (",var2,"%)", sep = "")) + 
#   stat_ellipse() +
#   coord_fixed() +
#   ggsave("figures/seedbank_pcoa.pdf", bg = "white", height = 8, width = 10)
# 

# seedbank.groups <- filter(design, sample.type == "RNA") %>% 
#   full_join(seedbank.traj.df) %>% 
#   mutate(month = month(date, label = T, abbr = T)) %>% 
#   select(month)
# seedbank.groups <- factor(seedbank.groups$month)
# sb.bdisp <- betadisper(seedbank.dist, group = seedbank.groups)
# plot(sb.bdisp)
# anova(sb.bdisp)
# boxplot(sb.bdisp)

active.pca <- rda(OTUs.REL.active)

# 
# OTUs.dist <- vegdist(OTUs.REL, method = "euclidean")
# OTUs.pcoa <- cmdscale(OTUs.dist, eig = T)
var1 <- round(eigenvals(active.pca)[1] / sum(eigenvals(active.pca)) * 100, 2)
var2 <- round(eigenvals(active.pca)[2] / sum(eigenvals(active.pca)) * 100, 2)

# OTUs.traj <- OTUs.pcoa$points
# colnames(OTUs.traj) <- c("Comp.1", "Comp.2")
# active.traj <- OTUs.traj[which(design$sample.type == "RNA"),]
# total.traj <- OTUs.traj[which(design$sample.type == "DNA"),]

active.traj.df <- cbind.data.frame(
  sample.id = design$sample.id[which(design$sample.type == "RNA")], 
  as.data.frame(scores(active.pca, display = "sites"))) 

active.spec.scores.pc <- as.data.frame(scores(active.pca, display = "species")) %>% rownames_to_column(var = "OTU") %>% 
  filter(PC1 > .2 | PC2 > .2) %>% 
  left_join(OTU.tax)
familynames <- stringr::str_split(active.spec.scores.pc$Family, "_")
fam.name.vec <- vector(length = length(active.spec.scores.pc$Family))
for(i in 1:length(active.spec.scores.pc$Family)){
  thisfamily <- familynames[[i]]
  fam.name.vec[i] <- paste(thisfamily[1], "sp.")
} 
active.spec.scores.pc$name <- fam.name.vec

scale.arrows0 <- 1.5

fig.pca <- full_join(design[which(design$sample.type == "RNA"),], active.traj.df) %>%
  ggplot(aes(x = PC1, y = PC2)) + 
  geom_hline(aes(yintercept = 0), alpha = 0.1) + 
  geom_vline(aes(xintercept = 0), alpha = 0.1) +
  #stat_ellipse(aes(color = lubridate::month(date, label = T)), alpha = 0.5) +
  geom_path(alpha = 0.3, arrow = arrow(angle = 15, length = unit(0.1, "inches"),
                  ends = "last", type = "closed")) + 
  geom_point(aes(color = lubridate::month(date, label = T),
                 shape = lubridate::month(date, label = T)), alpha = 0.5, size = 3) +
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) +
  scale_shape_manual(values = rep(15:18,3)) +
  # geom_segment(data = active.spec.scores.pc,
  #              aes(x = 0, y = 0,
  #                  xend = scale.arrows0*PC1,
  #                  yend = scale.arrows0*PC2),
  #              alpha = 0.7,
  #              size = 1,
  #              color = "black",
  #              arrow = arrow(
  #                angle = 20,
  #                length = unit(.1, "inches"),
  #                type = "open"),
  #            show.legend = F) +
  # geom_text_repel(data = active.spec.scores.pc,
  #                  aes(x = scale.arrows0*PC1,
  #                      y = scale.arrows0*PC2),
  #                  label = active.spec.scores.pc$name,
  #                  # nudge_x = c(-.2, .4, .5, .4),
  #                  nudge_y = c(.1, 0.05, .05, 0),
  #                  point.padding = .5, alpha = 1) +
  xlab(paste("PC 1 (",var1,"%)", sep = "")) + 
  ylab(paste("PC 2 (",var2,"%)", sep = "")) +
  coord_fixed() +
  ggsave("figures/fig1a.pdf", bg = "white", height = 6, width = 8)
fig.pca

```

### Explaining temporal beta diversity
#### DDRs
```{r}
# 
# samplingdays <- env.data$date
# time.dist <- matrix(NA, nrow = length(samplingdays), ncol = length(samplingdays))
# for(i in 1:length(samplingdays)){
#   for(j in 1:length(samplingdays)){
#     time.dist[i,j] <- (samplingdays[i] - samplingdays[j])
#   }
# }
# time.dist <- as.dist(time.dist)
# 
# env.vars <- env.data[c("temp", "spc", "oxygen", "secchi", "ph", "tp", "tn", "doc")]
# env.dist <- cluster::daisy(scale(env.vars), metric = "gower")
# 
# 
# OTUs.betapart <- beta.pair(x = OTUs.total.pa, index.family = "jaccard")
# OTUs.nest <- OTUs.betapart$beta.jne
# OTUs.turn <- OTUs.betapart$beta.jtu
# OTUs.jac <- OTUs.betapart$beta.jac
# 
# OTUs.act.betapart <- beta.pair(x = OTUs.active.pa, index.family = "jaccard")
# OTUs.act.nest <- OTUs.act.betapart$beta.jne
# OTUs.act.turn <- OTUs.act.betapart$beta.jtu
# OTUs.act.jac <- OTUs.act.betapart$beta.jac
# 
# 
# timeddr <- data.frame(total_turn = as.vector(OTUs.turn),
#                       total_nest = as.vector(OTUs.nest),
#                       total_jac = as.vector(OTUs.jac),
#                       active_turn = as.vector(OTUs.act.turn),
#                       active_nest = as.vector(OTUs.act.nest),
#                       active_jac = as.vector(OTUs.act.jac),
#                       EnvDist = as.vector(env.dist), 
#                       Time = as.vector(time.dist))
# 
# betapart.ddr.env <- timeddr %>% gather(key = Community, value = Dissimilarity, -EnvDist, -Time) %>% 
#   ggplot(aes(x = EnvDist, y = Dissimilarity, color = Community)) + 
#   geom_jitter(alpha = 0.1, size = 0.7) +
#   geom_smooth(size = 2, method = "lm") +
#   scale_y_continuous(limits = c(0,1))
# 
# betapart.ddr.time <- timeddr %>% gather(key = Community, value = Dissimilarity, -EnvDist, -Time) %>% 
#   ggplot(aes(x = Time, y = Dissimilarity, color = Community)) + 
#   geom_jitter(alpha = 0.1, size = 0.7) +
#   geom_smooth(size = 2) +
#   ylab("Community Dissimilarity\n") + 
#   xlab("Time lag (days)") +
#   scale_y_continuous(limits = c(0,1))
# 
# plot_grid(betapart.ddr.time + theme(legend.position = "none"), 
#           betapart.ddr.env, ncol = 2, align = 'hv', axis = 'l',
#           labels = c("A", "B")) +
#   ggsave(filename = "figures/betapart_ddrs.pdf", width = 10, height = 6, units = "in", bg = "white")
```
#### RDA
```{r}
#### Constrained Ordination
samplingdays <- env.data$date
intervals <- as.vector(diff.Date(samplingdays))
w <- 1/(intervals/max(intervals))
aem.out <- adespatial::aem.time(n = length(samplingdays), w = w, moran = TRUE)
aem.eigs <- data.frame(aem.out$aem)
aem.out$values
colnames(aem.eigs) <- paste0("AEM",seq(1,ncol(aem.eigs)))

env.preds <- na.omit(env.data[c("sample.id", "temp", "spc", "oxygen", "secchi", "ph", "tp", "tn", "doc")])

OTUs.rel.act.subs <- OTUs.REL.active[design$sample.id[which(design$sample.type == "RNA")] %in% env.preds$sample.id,]
aem.subs <- aem.eigs[design$sample.id[which(design$sample.type == "RNA")] %in% env.preds$sample.id,]

# #active.varpar <- varpart(OTUs.rel.act.subs, env.preds)
active.rda.0 <- rda(OTUs.rel.act.subs ~ 1, data = aem.subs)
# active.rda.1 <- rda(OTUs.rel.act.subs ~ ., data = aem.subs)
# active.rda.2 <- rda(OTUs.rel.act.subs ~ AEM2 + AEM4 + AEM5 + AEM6, data = aem.subs)
# active.rda <- ordiR2step(active.rda.0, scope = formula(active.rda.2), direction = "forward")
# 
# active.rda.env.0 <- rda(OTUs.rel.act.subs ~ 1, data = env.preds)
# active.rda.env.1 <- rda(OTUs.rel.act.subs ~ ., data = env.preds)
# active.rda.env <- ordiR2step(active.rda.env.0, scope = formula(active.rda.env.1), direction = "forward")

envfit(active.rda.0, aem.subs)

time.mod <- model.matrix(~ -1 + AEM1 + AEM2 + AEM3 + AEM4 + AEM5 + AEM6 + AEM7 + AEM15, data = aem.subs)
env.mod <- model.matrix(~ temp + doc + oxygen + spc + ph + secchi, data = env.preds)

# RDA without controlling for time
active.rda <- rda(OTUs.rel.act.subs ~ env.mod)

active.rda.sum <- summary(active.rda)

env.vecs <- cbind.data.frame(center = 0, active.rda.sum$biplot[,c(1:2)])
rownames(env.vecs) <- c("Temp", "DOC", "DO", "SpC", "pH", "secchi")
spec.scores <- cbind.data.frame(center = 0, scores(active.rda)$species)
site.scores <- as.data.frame(scores(active.rda)$sites)


active.site.scores <- site.scores %>% rownames_to_column(var = "sample.name") %>% inner_join(design)
active.spec.scores <- spec.scores[which(abs(spec.scores$RDA1) > 0.2 | abs(spec.scores$RDA2) > 0.2),] %>% 
  rownames_to_column(var = "OTU") %>% inner_join(OTU.tax)

# active.spec.scores.rda <- as.data.frame(scores(active.rda, display = "species")) %>% rownames_to_column(var = "OTU") 
#   filter(RDA1 > .2 | RDA2 > .2) %>% 
#   left_join(OTU.tax)
familynames <- stringr::str_split(active.spec.scores$Family, "_")
fam.name.vec <- vector(length = length(active.spec.scores$Family))
for(i in 1:length(active.spec.scores$Family)){
  thisfamily <- familynames[[i]]
  fam.name.vec[i] <- paste(thisfamily[1], "sp.")
} 
active.spec.scores$name <- fam.name.vec
active.spec.scores$name <- fam.name.vec

var.explain1 <- round(eigenvals(active.rda)[1]/sum(eigenvals(active.rda))*100,2)
var.explain2 <- round(eigenvals(active.rda)[2]/sum(eigenvals(active.rda))*100,2)

scale.arrows <- 1.4
scale.arrows2 <- 1
fig.rda <- 
  ggplot() + 
  # Add site scores as points
  # geom_point(data = site.scores, aes(x = RDA1, y = RDA2)) + 
  geom_hline(aes(yintercept = 0), alpha = 0.1) + 
  geom_vline(aes(xintercept = 0), alpha = 0.1) +
  
  geom_point(data = active.site.scores, alpha = 0.5, size = 3,
            aes(x = RDA1, y = RDA2, shape = lubridate::month(date, label = T, abbr = T),
                color = lubridate::month(date, label = T, abbr = T))) + 
  scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.2), 9)) +
  scale_shape_manual(values = rep(15:18,3)) +
  # Add species scores as vectors
  geom_segment(data = active.spec.scores, 
               aes(x = center, y = center, 
                   xend = scale.arrows*RDA1, yend = scale.arrows*RDA2), 
               alpha = .5, color = "black",
             arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"), 
             show.legend = F) + 
  geom_text_repel(data = active.spec.scores, 
               aes(x = scale.arrows*RDA1, y = scale.arrows*RDA2), 
               label = active.spec.scores$name, alpha = 1, point.padding = .5) + 

  # Add environmental vectors in blue
  geom_segment(data = env.vecs,
             aes(x = center, y = center, 
                 xend = scale.arrows2*RDA1, 
                 yend = scale.arrows2*RDA2),
             arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"),
             show.legend = F, alpha = .5, color = "blue") +
  
  geom_text(data = env.vecs, 
                  aes(x = (scale.arrows2+.1)*RDA1, y = (scale.arrows2+.1)*RDA2),
                    label = rownames(env.vecs), color = "blue", nudge_y = c(.1,0,0,0,0,0), nudge_x = c(.2,0,0,0,0,0),
                   segment.alpha = 0.3, alpha = 1, point.padding = .5) +
  
   
  coord_fixed() + 
  scale_x_reverse() +
  labs(x = paste0("RDA 1 (",var.explain1,"%)"),
       y = paste0("RDA 2 (",var.explain2,"%)")) +
  ggsave("figures/fig1b1.pdf", bg = "white", height = 6, width = 8)
fig.rda

# RDA controlling for time
# active.rda <- rda(OTUs.rel.act.subs ~ env.mod + Condition(time.mod))
# 
# active.rda.sum <- summary(active.rda)
# 
# env.vecs <- cbind.data.frame(center = 0, active.rda.sum$biplot[,c(1:2)])
# rownames(env.vecs) <- c("Temp", "DOC", "DO", "SpC", "pH", "secchi")
# spec.scores <- cbind.data.frame(center = 0, scores(active.rda)$species)
# site.scores <- as.data.frame(scores(active.rda)$sites)
# 
# 
# active.site.scores <- site.scores %>% rownames_to_column(var = "sample.name") %>% inner_join(design)
# active.spec.scores <- spec.scores[which(abs(spec.scores$RDA1) > 0.06 | abs(spec.scores$RDA2) > 0.06),] %>% 
#   rownames_to_column(var = "OTU") %>% inner_join(OTU.tax)
# 
# # active.spec.scores.rda <- as.data.frame(scores(active.rda, display = "species")) %>% rownames_to_column(var = "OTU") 
# #   filter(RDA1 > .2 | RDA2 > .2) %>% 
# #   left_join(OTU.tax)
# familynames <- stringr::str_split(active.spec.scores$Family, "_")
# fam.name.vec <- vector(length = length(active.spec.scores$Family))
# for(i in 1:length(active.spec.scores$Family)){
#   thisfamily <- familynames[[i]]
#   fam.name.vec[i] <- paste(thisfamily[1], "sp.")
# } 
# active.spec.scores$name <- fam.name.vec
# active.spec.scores$name <- fam.name.vec
# 
# var.explain1 <- round(eigenvals(active.rda)[1]/sum(eigenvals(active.rda))*100,2)
# var.explain2 <- round(eigenvals(active.rda)[2]/sum(eigenvals(active.rda))*100,2)
# 
# scale.arrows <- 6
# scale.arrows2 <- 1.5
# fig.rda <- ggplot() + 
#   # Add site scores as points
#   # geom_point(data = site.scores, aes(x = RDA1, y = RDA2)) + 
#   geom_point(data = active.site.scores, alpha = 0.9, size = 3,
#             aes(x = RDA1, y = RDA2, 
#                 color = lubridate::month(date, label = T, abbr = T),
#                 shape = lubridate::month(date, label = T, abbr = T))) + 
#   scale_color_manual(values = shifter(darken(plasma(12, direction = -1),.1), 9)) +
#   scale_shape_manual(values = c(0:2,5:14)) +
#   # Add species scores as vectors
#   geom_segment(data = active.spec.scores, 
#                aes(x = center, y = center, 
#                    xend = scale.arrows*RDA1, yend = scale.arrows*RDA2), 
#                alpha = .5, color = "black",
#              arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"), 
#              show.legend = F) + 
#   geom_text_repel(data = active.spec.scores, 
#                aes(x = scale.arrows*RDA1, y = scale.arrows*RDA2), 
#                label = active.spec.scores$name, alpha = 1, point.padding = .3) + 
# 
#   # Add environmental vectors in blue
#   geom_segment(data = env.vecs,
#              aes(x = center, y = center, 
#                  xend = scale.arrows2*RDA1, 
#                  yend = scale.arrows2*RDA2),
#              arrow = arrow(angle = 20, length = unit(.1, "inches"), type = "open"),
#              show.legend = F, alpha = .5, color = "blue") +
#   geom_text_repel(data = env.vecs, 
#                   aes(x = scale.arrows2*RDA1, y = scale.arrows2*RDA2),
#                     label = rownames(env.vecs), color = "blue", 
#                    segment.alpha = 0.3, alpha = 1) + 
#   coord_fixed() + 
#   labs(x = paste0("RDA 1 (",var.explain1,"%)"),
#        y = paste0("RDA 2 (",var.explain2,"%)")) +
#   ggsave("figures/fig1b2.pdf", bg = "white", height = 6, width = 8)

# plot combined figure
plot_grid(fig.pca, fig.rda + theme(legend.position = "none"), align = "hv", axis = "lr", ncol = 2, labels = "AUTO") +
  ggsave("figures/fig1.pdf", width = 12, height = .6*4/4*12)

## so this plot shows, after controlling for temporal autocorrelation, which environmental drivers are most important for some taxa
```

# Is negative frequency dependence disproportionately stronger on rare than common taxa?
```{r}
# calc.neg.freq.dep <- function(sbs){
#   y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
#   for(i in 1:(nrow(sbs)-1)){
#     y.s[i,] = log(sbs[i+1,] / sbs[i,])
#   }
#   eq.freq <- vector(length = ncol(y.s))
#   fd <- vector(length = ncol(y.s))
#   sbs.rel <- decostand(sbs[-nrow(sbs),], method = "total")
#   for(i in 1:ncol(y.s)){
#     mod <- lm(y.s[,i] ~ sbs.rel[,i])
#     eq.freq[i] <- -coef(mod)[1] / coef(mod)[2]
#     fd[i] <- coef(mod)[2]
#   }
#   # plot(log10(eq.freq), log10(-fd), 
#   #      ylab = "log10 Negative Freq. Dependence",
#   #      xlab = "log10 Equilibrium Frequency")
#   return(na.omit(cbind.data.frame(eq.freq, fd)))
# }

# using mean rel abund
calc.neg.freq.dep <- function(sbs){
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  eq.freq <- vector(length = ncol(y.s))
  fd <- vector(length = ncol(y.s))
  sbs.rel <- decostand(sbs[-nrow(sbs),], method = "total")
  for(i in 1:ncol(y.s)){
    mod <- lm(y.s[,i] ~ sbs.rel[,i])
    eq.freq[i] <- mean(sbs.rel[,i])
    fd[i] <- coef(mod)[2]
  }
  # plot(log10(eq.freq), log10(-fd), 
  #      ylab = "log10 Negative Freq. Dependence",
  #      xlab = "log10 Equilibrium Frequency")
  return(na.omit(cbind.data.frame(eq.freq, fd)))
}

shuffle.ts <- function(species.ts){
  species.ts[sample(1:length(species.ts))]
}

get.null.distr <- function(comm, iters = 5000){
  pb <- progress_bar$new(total = iters)
  nfd.cov <- rep(NA, iters)
  for(i in 1:iters){
    nullcom <- apply(comm, MARGIN = 2, FUN = shuffle.ts) # randomize sampling order
    nfd <- calc.neg.freq.dep(nullcom)
    nfd <- nfd %>% filter(-fd > 0)
    nfd.cov[i] <- cov(log(nfd$eq.freq), log(-nfd$fd))
    pb$tick()
  }
  return(nfd.cov)
}

# OTUs.ranked <- OTUs.total[,names(sort(colSums(OTUs.total), decreasing = T))]
env.data %>% ggplot(aes(x = date, y = temp)) + 
  geom_point() + 
  geom_line() +
  scale_x_date(breaks = "month") +
  coord_flip()
# separate OTUs present across the nearly whole time series
OTUs.total.persistent <- OTUs.total[,which(
  colMeans(decostand(OTUs.total, method = "pa")) >= .8)]
OTUs.active.persistent <- OTUs.active[,colnames(OTUs.total.persistent)]

# How many taxa to consider?
ncol(OTUs.total.persistent)


OTUs.total.nozero <- replace(OTUs.total.persistent, which(OTUs.total.persistent == 0), .5)
nfd.total <- calc.neg.freq.dep(OTUs.total.nozero)
nfd.total %>% filter(-fd > 0) -> nfd.total
nfd.total.cov <- cov(log(nfd.total$eq.freq), log(-nfd.total$fd))
nfd.total %>% ggplot(aes(x = eq.freq, y = -fd)) +
  geom_point(alpha = 0.5) + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm") + 
  ylab("Negative Frequency Dependence") + 
  xlab("Equilibrium Frequency") +
  ggsave("figures/NFD_total.pdf", width = 6, height = 3/4*6)

OTUs.active.nozero <- replace(OTUs.active.persistent, which(OTUs.active.persistent == 0), .5)
nfd.active <- calc.neg.freq.dep(OTUs.active.nozero)
nfd.active %>% filter(-fd > 0) -> nfd.active
nfd.active.cov <- cov(log(nfd.active$eq.freq), log(-nfd.active$fd))
nfd.active %>% ggplot(aes(x = eq.freq, y = -fd)) +
  geom_point(alpha = 0.5) + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm") + 
  ylab("Negative Frequency Dependence") + 
  xlab("Equilibrium Frequency") +
  ggsave("figures/NFD_active.pdf", width = 6, height = 4.5)

rbind.data.frame(cbind.data.frame(nfd.active, subset = "Active"),
                 cbind.data.frame(nfd.total, subset = "Total")) %>% 
  ggplot(aes(x = eq.freq, y = -fd, color = subset, fill = subset)) +
  geom_point(alpha = 0.5) + 
  scale_y_log10() + 
  scale_x_log10() +
  geom_smooth(method = "lm") + 
  ylab("Negative Frequency Dependence") + 
  xlab("Equilibrium Frequency") +
  ggsave("figures/NFD_both.pdf", width = 6, height = 4.5)
 
cbind.data.frame(active_freq = nfd.active$eq.freq,
                 active_nfd = nfd.active$fd, 
                 total_freq = nfd.total$eq.freq,
                 total_nfd = nfd.total$fd) %>% 
  mutate(freq_diff = active_freq - total_freq,
         nfd_diff = active_nfd - total_nfd) %>% 
  ggplot(aes(x = freq_diff, y = nfd_diff)) + 
  geom_hline(aes(yintercept = 0), alpha = 0.2) + 
  geom_vline(aes(xintercept = 0), alpha = 0.2) +
  geom_point(alpha = 0.5) +
  annotate("text", x = .06, y = 200, label = "Higher Eq. freq. and\n stronger NFD in active") +
  annotate("text", -.04, y = -400, label = "Higher Eq. freq. \n and stronger NFD \n in total") +
  labs(x = "Diff. in equilibrium frequency \n (active - total)",
       y = "Diff. in NFD \n (active - total)") +
  ggsave("figures/NFD_comparison.pdf", width = 5, height = 5)

# Generate null distributions
iternum <- 5000
tot.nd <- get.null.distr(comm = OTUs.total.nozero, iters = iternum)
act.nd <- get.null.distr(comm = OTUs.active.nozero, iters = iternum)

# Compare observed with null
hist(tot.nd, breaks = 20); abline(v = nfd.total.cov, col = "red")
(tot.nfd.pval <- rank(c(nfd.total.cov, tot.nd))[1] / length(c(nfd.total.cov, tot.nd)))
hist(act.nd, breaks = 20); abline(v = nfd.active.cov, col = "red")
(act.nfd.pval <- rank(c(nfd.active.cov, act.nd))[1] / length(c(nfd.active.cov, act.nd)))
# hist(sb.nd, breaks = 20); abline(v = nfd.sb.cov, col = "red")
# (sb.nfd.pval <- rank(c(nfd.sb.cov, sb.nd))[1] / length(c(nfd.sb.cov, sb.nd)))

# nfd.obs <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = nfd.total.cov),
#                  cbind.data.frame(subset = "Active", nfd = nfd.active.cov),
#                  cbind.data.frame(subset = "Seed Bank", nfd = nfd.sb.cov)) 
# 
# nfd.nulldists <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = tot.nd),
#                  cbind.data.frame(subset = "Active", nfd = act.nd),
#                  cbind.data.frame(subset = "Seed Bank", nfd = sb.nd)) 

nfd.obs <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = nfd.total.cov),
                 cbind.data.frame(subset = "Active", nfd = nfd.active.cov)) 

nfd.nulldists <- rbind.data.frame(cbind.data.frame(subset = "Total", nfd = tot.nd),
                 cbind.data.frame(subset = "Active", nfd = act.nd)) 

nfd.nulldists %>% 
  ggplot(aes(y = nfd, x = subset)) + 
  #geom_jitter(color = "grey", alpha = 0.1) + 
  geom_violin(alpha = 0.5, show.legend = F, fill = "grey") + 
  geom_point(data = nfd.obs, aes(x = subset, y = nfd), 
             color = "black", size = 5, show.legend = F) + 
  scale_fill_manual(values = my.cols[c(1,4)]) +
  coord_flip() +
  ylab("Covariance between Equilibrium Frequency\n and Negative Frequency Dependence") + 
  xlab("") + 
  ggsave("figures/negative_freq_depend.pdf", bg = "white", width = 6, height = 4, units = "in") +
  ggsave("figures/negative_freq_depend.png", bg = "white", width = 6, height = 4, units = "in", dpi = 600)

# Calculate standardized effect sizes
nfd.summary <- nfd.nulldists %>% 
  group_by(subset) %>% summarize(mean.nfd = mean(nfd), sd.nfd = sd(nfd)) %>% 
  left_join(nfd.obs) %>% mutate(nfd.ses = (nfd - mean.nfd)/sd.nfd) %>% 
  mutate(nfd.strength = nfd / mean.nfd)
nfd.summary
```

```{r}
# env.data %>% ggplot(aes(x = date, y = temp)) +
#   geom_point() +
#   geom_line() +
#   scale_x_date(breaks = "month") +
#   coord_flip()
# 
# # separate OTUs present across the nearly whole time series
# index.summer.2013 <- design %>%
#   filter(sample.type == "DNA") %>%
#   filter(month(date) >= 5 & month(date) <=10) %>%
#   filter(year(date) == 2013) %>% .$sample.name
# 
# index.summer.2014 <- design %>%
#   filter(sample.type == "DNA") %>%
#   filter(month(date) >= 5 & month(date) <=10) %>%
#   filter(year(date) == 2014) %>% .$sample.name
# 
# index.summer.2015 <- design %>%
#   filter(sample.type == "DNA") %>%
#   filter(month(date) >= 5 & month(date) <=10) %>%
#   filter(year(date) == 2015) %>% .$sample.name
# 
# nfd.obs.summer <- data.frame(year = NA, subset = NA, nfd = NA)
# nfd.nulldists.summer <- data.frame(year = NA, subset = NA, nfd = NA)
# 
# for(y in c(2013, 2014, 2015)){
#   index = paste0("index.summer.", y)
#   OTUs.total.persistent.summer <- OTUs.total[eval(parse(text = index)), which(colMeans(decostand(OTUs.total, method = "pa")) >= .8)]
#   OTUs.active.persistent.summer <- OTUs.active[,colnames(OTUs.total.persistent.summer)]
#   
#   OTUs.total.nozero.summer <- replace(OTUs.total.persistent.summer, which(OTUs.total.persistent.summer == 0), .5)
#   nfd.total.summer <- calc.neg.freq.dep(OTUs.total.nozero.summer)
#   nfd.total.summer %>% filter(-fd > 0) -> nfd.total.summer
#   nfd.total.cov.summer <- cov(log(nfd.total.summer$eq.freq), log(-nfd.total.summer$fd))
#   nfd.total.summer %>% ggplot(aes(x = eq.freq, y = -fd)) +
#     geom_point() +
#     scale_y_log10() +
#     scale_x_log10() +
#     geom_smooth(method = "lm") +
#     ylab("Negative Frequency Dependence") +
#     xlab("Equilibrium Frequency")
#   
#   OTUs.active.nozero.summer <- replace(OTUs.active.persistent.summer, which(OTUs.active.persistent.summer == 0), .5)
#   nfd.active.summer <- calc.neg.freq.dep(OTUs.active.nozero.summer)
#   nfd.active.summer %>% filter(-fd > 0) -> nfd.active.summer
#   nfd.active.cov.summer <- cov(log(nfd.active.summer$eq.freq), log(-nfd.active.summer$fd))
#   nfd.active.summer %>% ggplot(aes(x = eq.freq, y = -fd)) +
#     geom_point() +
#     scale_y_log10() +
#     scale_x_log10() +
#     geom_smooth(method = "lm") +
#     ylab("Negative Frequency Dependence") +
#     xlab("Equilibrium Frequency")
#   
#   
#   # Generate null distributions
#   iternum <- 5000
#   tot.nd.summer <- get.null.distr(comm = OTUs.total.nozero.summer, iters = iternum)
#   act.nd.summer <- get.null.distr(comm = OTUs.active.nozero.summer, iters = iternum)
#   
#   # Compare observed with null
#   hist(tot.nd.summer, breaks = 20); abline(v = nfd.total.cov.summer, col = "red")
#   (tot.nfd.pval.summer <- rank(c(nfd.total.cov.summer, tot.nd.summer))[1] / length(c(nfd.total.cov.summer, tot.nd.summer)))
#   hist(act.nd.summer, breaks = 20); abline(v = nfd.active.cov.summer, col = "red")
#   (act.nfd.pval.summer <- rank(c(nfd.active.cov.summer, act.nd.summer))[1] / length(c(nfd.active.cov.summer, act.nd.summer)))
#   
#   nfd.obs.summer <- rbind.data.frame(
#     nfd.obs.summer,
#     cbind.data.frame(year = y, subset = "Total", nfd = nfd.total.cov.summer),
#     cbind.data.frame(year = y, subset = "Active", nfd = nfd.active.cov.summer))
#   
#   nfd.nulldists.summer <- rbind.data.frame(
#     nfd.nulldists.summer,
#     cbind.data.frame(year = y, subset = "Total", nfd = tot.nd.summer),
#     cbind.data.frame(year = y, subset = "Active", nfd = act.nd.summer))
# }
# nfd.obs.summer <- na.omit(nfd.obs.summer)
# nfd.nulldists.summer <- na.omit(nfd.nulldists.summer)

#saveRDS(nfd.obs.summer, "analysis/temp-results/nfd_obs_summer.rda")
#saveRDS(nfd.nulldists.summer, "analysis/temp-results/nfd_nulldists_summer.rda")
nfd.obs.summer <- readRDS("analysis/temp-results/nfd_obs_summer.rda")
nfd.nulldists.summer <- readRDS("analysis/temp-results/nfd_nulldists_summer.rda")


nfd.nulldists.summer %>%
  ggplot(aes(y = nfd, x = subset)) +
  facet_wrap(~ year, ncol = 1) +
  #geom_jitter(color = "grey", alpha = 0.1) +
  geom_violin(alpha = 0.5, show.legend = F, fill = "grey") +
  geom_point(data = nfd.obs.summer, aes(x = subset, y = nfd),
             color = "black", size = 5, show.legend = F) +
  scale_fill_manual(values = my.cols[c(1,4)]) +
  coord_flip() +
  ylab("Covariance between Equilibrium Frequency\n and Negative Frequency Dependence") +
  xlab("")

labs <- nfd.nulldists.summer %>% 
  group_by(year, subset) %>% 
  summarize(mean_nfd = mean(nfd))

nfd.nulldists.summer %>% 
  ggplot(aes(x = nfd, fill = subset)) +
  facet_wrap(~year, ncol = 1) +
  geom_histogram(alpha = 0.5, bins = 100) +
  geom_vline(data = nfd.obs.summer, aes(xintercept = nfd, color = subset), size = 1) +
  theme(legend.position = "none") +
  labs(x = "Cov(EQ frequency, NFD)",
       y = "") +
  geom_text(data = labs, aes(x = mean_nfd, y = 100, color = subset, label = subset), color = "black") +
  ggsave("figures/nfd_by_year_summer.pdf", width = 4, height = 9)

# Calculate standardized effect sizes
nfd.summary.summer <- nfd.nulldists.summer %>%
  group_by(subset, year) %>% summarize(mean.nfd = mean(nfd), sd.nfd = sd(nfd)) %>%
  left_join(nfd.obs.summer) %>% mutate(nfd.ses = (nfd - mean.nfd)/sd.nfd) %>%
  mutate(nfd.strength = nfd / mean.nfd)
nfd.summary.summer


(pval.2013.tot <- rank(c(filter(nfd.obs.summer, subset == "Total", year == 2013)$nfd, 
       filter(nfd.nulldists.summer, subset == "Total", year == 2013)$nfd))[1] / 5001)
(pval.2013.act <- rank(c(filter(nfd.obs.summer, subset == "Active", year == 2013)$nfd, 
       filter(nfd.nulldists.summer, subset == "Active", year == 2013)$nfd))[1] / 5001)
(pval.2014.tot <- rank(c(filter(nfd.obs.summer, subset == "Total", year == 2014)$nfd, 
       filter(nfd.nulldists.summer, subset == "Total", year == 2014)$nfd))[1] / 5001)
(pval.2014.act <- rank(c(filter(nfd.obs.summer, subset == "Active", year == 2014)$nfd, 
       filter(nfd.nulldists.summer, subset == "Active", year == 2014)$nfd))[1] / 5001)
(pval.2015.tot <- rank(c(filter(nfd.obs.summer, subset == "Total", year == 2015)$nfd, 
       filter(nfd.nulldists.summer, subset == "Total", year == 2015)$nfd))[1] / 5001)
(pval.2015.act <- rank(c(filter(nfd.obs.summer, subset == "Active", year == 2015)$nfd, 
       filter(nfd.nulldists.summer, subset == "Active", year == 2015)$nfd))[1] / 5001)
```

```{r}
# separate OTUs present across the nearly whole time series
# index.winter.2013 <- design %>%
#   filter(sample.type == "DNA") %>%
#   filter(month(date) <= 5 | month(date) >= 10) %>%
#   filter(year(date) == 2013) %>% .$sample.name
# 
# index.winter.2014 <- design %>%
#   filter(sample.type == "DNA") %>%
#   filter(month(date) <= 5 | month(date) >= 10) %>%
#   filter(year(date) == 2014) %>% .$sample.name
# 
# index.winter.2015 <- design %>%
#   filter(sample.type == "DNA") %>%
#   filter(month(date) <= 5 | month(date) >= 10) %>%
#   filter(year(date) == 2015) %>% .$sample.name
# 
# nfd.obs.winter <- data.frame(year = NA, subset = NA, nfd = NA)
# nfd.nulldists.winter <- data.frame(year = NA, subset = NA, nfd = NA)
# 
# for(y in c(2013, 2014, 2015)){
#   index = paste0("index.winter.", y)
#   OTUs.total.persistent.winter <- OTUs.total[eval(parse(text = index)), which(colMeans(decostand(OTUs.total, method = "pa")) >= .8)]
#   OTUs.active.persistent.winter <- OTUs.active[,colnames(OTUs.total.persistent.winter)]
#   
#   OTUs.total.nozero.winter <- replace(OTUs.total.persistent.winter, which(OTUs.total.persistent.winter == 0), .5)
#   nfd.total.winter <- calc.neg.freq.dep(OTUs.total.nozero.winter)
#   nfd.total.winter %>% filter(-fd > 0) -> nfd.total.winter
#   nfd.total.cov.winter <- cov(log(nfd.total.winter$eq.freq), log(-nfd.total.winter$fd))
#   nfd.total.winter %>% ggplot(aes(x = eq.freq, y = -fd)) +
#     geom_point() +
#     scale_y_log10() +
#     scale_x_log10() +
#     geom_smooth(method = "lm") +
#     ylab("Negative Frequency Dependence") +
#     xlab("Equilibrium Frequency")
#   
#   OTUs.active.nozero.winter <- replace(OTUs.active.persistent.winter, which(OTUs.active.persistent.winter == 0), .5)
#   nfd.active.winter <- calc.neg.freq.dep(OTUs.active.nozero.winter)
#   nfd.active.winter %>% filter(-fd > 0) -> nfd.active.winter
#   nfd.active.cov.winter <- cov(log(nfd.active.winter$eq.freq), log(-nfd.active.winter$fd))
#   nfd.active.winter %>% ggplot(aes(x = eq.freq, y = -fd)) +
#     geom_point() +
#     scale_y_log10() +
#     scale_x_log10() +
#     geom_smooth(method = "lm") +
#     ylab("Negative Frequency Dependence") +
#     xlab("Equilibrium Frequency")
#   
#   
#   # Generate null distributions
#   iternum <- 5000
#   tot.nd.winter <- get.null.distr(comm = OTUs.total.nozero.winter, iters = iternum)
#   act.nd.winter <- get.null.distr(comm = OTUs.active.nozero.winter, iters = iternum)
#   
#   # Compare observed with null
#   hist(tot.nd.winter, breaks = 20); abline(v = nfd.total.cov.winter, col = "red")
#   (tot.nfd.pval.winter <- rank(c(nfd.total.cov.winter, tot.nd.winter))[1] / length(c(nfd.total.cov.winter, tot.nd.winter)))
#   hist(act.nd.winter, breaks = 20); abline(v = nfd.active.cov.winter, col = "red")
#   (act.nfd.pval.winter <- rank(c(nfd.active.cov.winter, act.nd.winter))[1] / length(c(nfd.active.cov.winter, act.nd.winter)))
#   
#   nfd.obs.winter <- rbind.data.frame(
#     nfd.obs.winter,
#     cbind.data.frame(year = y, subset = "Total", nfd = nfd.total.cov.winter),
#     cbind.data.frame(year = y, subset = "Active", nfd = nfd.active.cov.winter))
#   
#   nfd.nulldists.winter <- rbind.data.frame(
#     nfd.nulldists.winter,
#     cbind.data.frame(year = y, subset = "Total", nfd = tot.nd.winter),
#     cbind.data.frame(year = y, subset = "Active", nfd = act.nd.winter))
# }
# nfd.obs.winter <- na.omit(nfd.obs.winter)
# nfd.nulldists.winter <- na.omit(nfd.nulldists.winter)

# saveRDS(nfd.obs.winter, "analysis/temp-results/nfd_obs_winter.rda")
# saveRDS(nfd.nulldists.winter, "analysis/temp-results/nfd_nulldists_winter.rda")
nfd.obs.winter <- readRDS("analysis/temp-results/nfd_obs_winter.rda")
nfd.nulldists.winter <- readRDS("analysis/temp-results/nfd_nulldists_winter.rda")


nfd.nulldists.winter %>%
  ggplot(aes(y = nfd, x = subset)) +
  facet_wrap(~ year, ncol = 1) +
  #geom_jitter(color = "grey", alpha = 0.1) +
  geom_violin(alpha = 0.5, show.legend = F, fill = "grey") +
  geom_point(data = nfd.obs.winter, aes(x = subset, y = nfd),
             color = "black", size = 5, show.legend = F) +
  scale_fill_manual(values = my.cols[c(1,4)]) +
  coord_flip() +
  ylab("Covariance between Equilibrium Frequency\n and Negative Frequency Dependence") +
  xlab("")

labs <- nfd.nulldists.winter %>% 
  group_by(year, subset) %>% 
  summarize(mean_nfd = mean(nfd))

nfd.nulldists.winter %>% 
  ggplot(aes(x = nfd, fill = subset)) +
  facet_wrap(~year, ncol = 1) +
  geom_histogram(alpha = 0.5, bins = 100) +
  geom_vline(data = nfd.obs.winter, aes(xintercept = nfd, color = subset), size = 1) +
  theme(legend.position = "none") +
  labs(x = "Cov(EQ frequency, NFD)",
       y = "") +
  geom_text(data = labs, aes(x = mean_nfd, y = 100, color = subset, label = subset), color = "black") +
  ggsave("figures/nfd_by_year_winter.pdf", width = 4, height = 9)

# Calculate standardized effect sizes
nfd.summary.winter <- nfd.nulldists.winter %>%
  group_by(subset, year) %>% summarize(mean.nfd = mean(nfd), sd.nfd = sd(nfd)) %>%
  left_join(nfd.obs.winter) %>% mutate(nfd.ses = (nfd - mean.nfd)/sd.nfd) %>%
  mutate(nfd.strength = nfd / mean.nfd)
nfd.summary.winter


(pval.2013.tot <- rank(c(filter(nfd.obs.winter, subset == "Total", year == 2013)$nfd, 
       filter(nfd.nulldists.winter, subset == "Total", year == 2013)$nfd))[1] / 5001)
(pval.2013.act <- rank(c(filter(nfd.obs.winter, subset == "Active", year == 2013)$nfd, 
       filter(nfd.nulldists.winter, subset == "Active", year == 2013)$nfd))[1] / 5001)
(pval.2014.tot <- rank(c(filter(nfd.obs.winter, subset == "Total", year == 2014)$nfd, 
       filter(nfd.nulldists.winter, subset == "Total", year == 2014)$nfd))[1] / 5001)
(pval.2014.act <- rank(c(filter(nfd.obs.winter, subset == "Active", year == 2014)$nfd, 
       filter(nfd.nulldists.winter, subset == "Active", year == 2014)$nfd))[1] / 5001)
(pval.2015.tot <- rank(c(filter(nfd.obs.winter, subset == "Total", year == 2015)$nfd, 
       filter(nfd.nulldists.winter, subset == "Total", year == 2015)$nfd))[1] / 5001)
(pval.2015.act <- rank(c(filter(nfd.obs.winter, subset == "Active", year == 2015)$nfd, 
       filter(nfd.nulldists.winter, subset == "Active", year == 2015)$nfd))[1] / 5001)
```


```{r}
# What is the effect of taxa number on response
# nfd.response <- function(OTUs = NULL, OTUs.active = NULL, num.seq = 10, num.iter = 999, plotit = F, incidence.freq = 0.9, max.taxa = NULL){
#   
#   # separate OTUs present across the whole time series
#   OTUs.ranked <- OTUs[,names(sort(colSums(OTUs), decreasing = T))]
#   OTUs.persistent <- OTUs.ranked[,which(
#     colMeans(decostand(OTUs.ranked, method = "pa")) >= incidence.freq)]
#   
#   # If we're looking at Active OTUs, make the switch here
#   if(!is.null(OTUs.active)){
#     if(is.null(OTUs)) print("please supply total otu table")
#     OTUs.persistent <- OTUs.active[,colnames(OTUs.persistent)]
#   }
#   
#   # decide how many values of cutoffs to test
#   max.taxa <- floor(seq(3, min(max.taxa, ncol(OTUs.persistent)), length.out = num.seq))
#   output <- data.frame(taxa = max.taxa)
#   
#   # loop through different otu thresholds
#   for(i in 1:length(max.taxa)){
#     if(max.taxa[i] > ncol(OTUs.persistent)) return(output)
#     otu.thresh <- min(max.taxa[i], ncol(OTUs.persistent))
#     OTUs.nozero <- OTUs.persistent[,1:otu.thresh] + 10^-10
#     nfd <- calc.neg.freq.dep(OTUs.nozero)
#     nfd <- nfd %>% filter(eq.freq > 10^-9) %>% filter(-fd > 0)
#     nfd.cov <- cov(log(nfd$eq.freq), log(-nfd$fd))
#     
#     if(plotit){
#       nfd %>% 
#         ggplot(aes(x = eq.freq, y = -fd)) +
#         geom_point() + 
#         scale_y_log10() + 
#         scale_x_log10() +
#         geom_smooth(method = "lm") + 
#         ylab("Negative Frequency Dependence") + 
#         xlab("Equilibrium Frequency")
#     }
#     
#     nd <- get.null.distr(comm = OTUs.nozero, iters = num.iter)
#     nfd.ses <- (nfd.cov - mean(nd)/sd(nd))
#     nfd.pval <- rank(c(nfd.cov, nd))[1] / length(c(nfd.cov, nd))
#     cov.strength <- nfd.cov / mean(nd)
#     se <- sd(nd)/sqrt(num.iter)
#     output[i,"p"] <- nfd.pval
#     output[i,"ses"] <- nfd.ses
#     output[i,"strength"] <- cov.strength
#     output[i,"stderr"] <- se
#     output[i,"mean"] <- mean(nd)
#     output[i,"nfdcov"] <- nfd.cov
#   }
#   
#   return(output)
# }
# 
# 
# active.nfd.response <- nfd.response(OTUs = OTUs.total, OTUs.active = OTUs.active, num.seq = 30, num.iter = 999, max.taxa = 100)
# total.nfd.response <- nfd.response(OTUs = OTUs.total, OTUs.active = NULL, num.seq = 30, num.iter = 999, max.taxa = 100)
# saveRDS(active.nfd.response, file = "analysis/temp-results/nfd_response_active.rda")
# saveRDS(total.nfd.response, file = "analysis/temp-results/nfd_response_total.rda")
active.nfd.response <- readRDS(file = "analysis/temp-results/nfd_response_active.rda")
total.nfd.response <- readRDS(file = "analysis/temp-results/nfd_response_total.rda")

nfd.response.df <- rbind.data.frame(
  cbind.data.frame(subset = "active", active.nfd.response),
  cbind.data.frame(subset = "total", total.nfd.response))
nfd.response.df %>% 
#  filter(p < 0.05) %>% 
  ggplot(aes(x = taxa, y = strength, color = subset)) +
  geom_hline(yintercept = 1) +
  geom_point() +
  geom_line() +
  labs(x = "Number of OTUs included in analysis",
       y = "Strength of NFD \n(observed / null mean)") +
  ggsave("figures/nfd_sensitivity.pdf", width = 6, height = 4)

cbind.data.frame(taxa = active.nfd.response$taxa, 
      strength = active.nfd.response$strength / total.nfd.response$strength) %>% 
  ggplot(aes(x = taxa, y = strength)) +
  geom_point() +
  geom_abline(slope = 0, intercept = 1, color = "gray") +
  geom_smooth(method = "loess", se = F) 
```

# Reappeareance Scores
```{r}
# Calculate a species 

reappearance.score <- function(DNA, RNA){
  # calculate # of times present but not active
  otu.diff <- decostand((DNA + RNA), method = "pa") - decostand(RNA, method = "pa")
  otu.diff.weighted <- otu.diff # * decostand(DNA, method = "total")
  return(colSums(otu.diff.weighted))
}

calc.score <- function(species){
  length()
}

(OTUs.total - OTUs.active)[,1]

reappearance.score <- function(DNA, RNA){
  
  
  otu.dna.pa <- decostand(DNA, method = "pa")
  otu.rna.pa <- decostand(RNA, method = "pa")
  otu.diff <- otu.dna.pa - otu.rna.pa
    
  otuscores <- vector(length = ncol(otu.diff))
  
  # for each OTU in the difference OTU matrix
  for(j in 1:ncol(otu.diff)){
    
    # first assign a score of 0
    otuscore <- 0
    
    # then, go row by row within each OTU
    for(i in 1:(nrow(otu.diff)-1)){
      
      # each time an OTU reappears, score it 1
      if(otu.rna.pa[i, j] == 0 & # absent from RNA at i
         otu.dna.pa[i, j] == 1 & # present in DNA at i
         otu.rna.pa[i+1, j] == 1){ # present again in RNA
        otuscore <- otuscore + 1
      }
    }
    otuscores[j] <- otuscore
    print(paste("completed otu #", j))
  }
  
  return(otuscores)
}

otu.reap <- reappearance.score(DNA = OTUs.total, RNA = OTUs.active)
otu.pres <- colSums(OTUs.total.pa)
otu.total.relabund <- colMeans(decostand((OTUs.total), method = "total"))
otu.active.relabund <- colMeans(decostand((OTUs.active), method = "total"))

persistent_taxa <- colnames(OTUs.total.persistent)

persistent_seedbank_ornot <- cbind.data.frame(OTU = colnames(OTUs.total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
  filter(times_present > 0) %>% 
  filter(relabund_total > 0) %>% 
  filter(times_present > 1) %>% 
  mutate(percent_transition = times_activated/times_present) %>% 
  mutate(seedbank = ifelse(percent_transition == 0, F,T)) %>% 
  mutate(persistent = ifelse(OTU %in% persistent_taxa, T,F))

# taxa not relying on seed bank 
no_sb_taxa <- persistent_seedbank_ornot %>% filter(persistent == T) %>% 
  filter(seedbank == F) %>% 
  left_join(OTU.tax)
yes_sb_taxa <- persistent_seedbank_ornot %>% filter(persistent == T) %>% 
  filter(seedbank == T) %>% 
  left_join(OTU.tax)


cbind.data.frame(OTU = colnames(OTUs.total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
  filter(times_present > 0) %>% 
  filter(relabund_total > 0) %>% 
  filter(times_present > 1) %>% 
  mutate(percent_transition = times_activated/times_present) %>% 
  mutate(seedbank = ifelse(percent_transition == 0, "Not reliant on seed bank", "Seed bank reliant")) %>% 
  mutate(persistent = ifelse(OTU %in% persistent_taxa, "Persistent", "Non-persistent")) %>% 
  ggplot(aes(y = times_activated, x = relabund_total, color = seedbank)) + 
  geom_jitter(width = 0.005, height = 0.005, alpha = 0.25) + 
  geom_smooth() + 
  facet_wrap(~persistent) +
  ylab("Number of activations from seed bank") +
  xlab("Average relative abundance in total community") +
  scale_x_log10() 
  #ggsave("figures/incidence_by_inactivity.pdf", bg = "white", width = 6, height = 6)

cbind.data.frame(OTU = colnames(OTUs.total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
  filter(times_present > 0) %>% 
  filter(relabund_total > 0) %>% 
  filter(relabund_active > 0) %>% 
  filter(times_present > 1) %>% 
  mutate(percent_transition = times_activated/times_present) %>% 
  mutate(seedbank = ifelse(percent_transition == 0, "Not reliant on seed bank", "Seed bank reliant")) %>% 
  mutate(persistent = ifelse(OTU %in% persistent_taxa, "Persistent", "Non-persistent")) %>% 
  ggplot(aes(y = times_activated, x = relabund_active, color = seedbank)) + 
  geom_jitter(width = 0.005, height = 0.005, alpha = 0.25) + 
  geom_smooth() + 
  facet_wrap(~persistent) +
  ylab("Number of activations from seed bank") +
  xlab("Average relative abundance in active community") +
  scale_x_log10()
  #ggsave("figures/incidence_by_inactivity.pdf", bg = "white", width = 6, height = 6)

cbind.data.frame(OTU = colnames(OTUs.total),
                 times_activated = otu.reap, 
                 times_present = otu.pres,
                 relabund_active = otu.active.relabund,
                 relabund_total = otu.total.relabund) %>%
  filter(times_present > 0) %>% 
  filter(relabund_total > 0) %>% 
  filter(relabund_active > 0) %>% 
  filter(times_present > 1) %>% 
  mutate(percent_transition = times_activated/times_present) %>% 
  mutate(seedbank = ifelse(percent_transition == 0, "Not reliant on seed bank", "Persistent via seed bank")) %>% 
  mutate(persistent = ifelse(OTU %in% persistent_taxa, "Persistent", "Non-persistent")) %>% 
  filter(persistent == "Persistent") %>% 
  gather(relabund_active, relabund_total, key = "subset", value = "relabund") %>% 
  mutate(subset = ifelse(subset == "relabund_active", "Active Portion", "Total Community")) %>% 
  filter(subset == "Active Portion") %>% 
  ggplot(aes(y = times_activated, x = relabund)) + 
  geom_jitter(size = 3, alpha = 0.5) + 
  geom_smooth(color = "black", span = 1) + 
  #facet_wrap(~subset, ncol = 1) +
  theme(legend.position = "bottom", legend.text = element_text(size = 10)) +
  ylab("Reactivations from seed bank") +
  xlab("Average relative abundance") +
  scale_x_log10() +
  ggsave("figures/fig5.pdf", width = 6, height = 3/4*6)
  
  

# pull out taxa who are present > 50% of time but inactive > 50% of time
# these may be potential seed bankers or sink populations form soils
cbind.data.frame(times_inactive = otu.reap, presences = otu.pres) %>%
  rownames_to_column(var = "OTU") %>% 
  mutate(prop_inactive = times_inactive/presences, 
         incidence = presences / (length(env.data$date))) %>% 
  filter(incidence > 0.9, prop_inactive > 0.1) -> seedbank.taxa
subset(OTU.tax, OTU %in% seedbank.taxa$OTU)
```

# Taxa Time Relationship
```{r}
# ttr.calc <- function(comm){
#   comm.pa <- decostand(comm, method = 'pa')
#   weeknum <- 1:nrow(comm.pa)
#   otusum <- numeric(nrow(comm.pa))
#   for(week in weeknum){
#     otusum[week] <- ifelse(week > 1, sum(colSums(comm.pa[1:week,]) > 0), 
#                            sum(comm.pa[1,]))
#   }
#   return(cbind.data.frame(week = weeknum, taxa = otusum))
# }
# 
# active.ttr <- ttr.calc(OTUs.active)
# total.ttr <- ttr.calc(OTUs.total)
# 
# ttrs <- rbind.data.frame(cbind.data.frame(
#   subset = "active", active.ttr),
#                  cbind.data.frame(subset = "total", total.ttr))
# ttrs %>% 
#   ggplot(aes(x = week, y = taxa, color = subset)) + 
#   geom_line(size = 2) + 
#   ggsave("figures/taxa_time_relationships.pdf", bg = "white", height = 8, width = 8)
```

# Conditionally Rare Taxa
```{r}
# identifyCRT <- function(spec.dyn){
#   n <- length(spec.dyn)
#   skew.numer <- sum((spec.dyn - mean(spec.dyn))^3)/n
#   skew.denom <- (sum((spec.dyn - mean(spec.dyn))^2)/n)^(3/2)
#   kurt.num <- sum((spec.dyn - mean(spec.dyn))^4)/n
#   kurt.denom <- (sum((spec.dyn - mean(spec.dyn))^2)/n)^2
#   skewness <- skew.numer / skew.denom
#   kurtosis <- kurt.num / kurt.denom
#   b <- (1 + skewness^2) / (kurtosis + 3)
#   return(b)
#  }
# 
# activeCRTs <- apply(X = OTUs.REL.active[,which(colMeans(OTUs.active.pa) > 0.5)], MARGIN = 2, FUN = identifyCRT)
# totalCRTs <- apply(X = OTUs.REL.total[,which(colMeans(OTUs.total.pa) > 0.5)], MARGIN = 2, FUN = identifyCRT)
# hist(activeCRTs)
# hist(totalCRTs)
# 
# 
# candidateCRTs <- names(activeCRTs[which(activeCRTs > 0.6)])
# candidateCRTs <- OTUs.REL.active[,candidateCRTs]
# candidateCRTs <- candidateCRTs[,apply(candidateCRTs, 2, max) > 0.01]
# matplot(candidateCRTs, type = 'l')
# subset(OTU.tax, OTU %in% colnames(candidateCRTs))
# as.data.frame(candidateCRTs) %>% 
#   rownames_to_column(var = "sample.name") %>% 
#   left_join(design) %>% 
#   gather(-sample.name, -sample.id, -sample.type, -date, 
#          key = "OTU", value = "abundance") %>% 
#   left_join(OTU.tax) %>% 
#   ggplot(aes(x = date, y = abundance, color = Genus)) + 
#   geom_line()

```


```{r}
# shuffle.ts <- function(species.ts){
#   species.ts[sample(1:length(species.ts))]
# }
# 
# get.null.mrs <- function(comm, iters = 100){
#   pb <- progress_bar$new(total = iters)
#   mrs.null <- rep(NA, iters)
#   for(i in 1:iters){
#     nullcom <- apply(comm, MARGIN = 2, FUN = shuffle.ts) # randomize sampling order
#     nullcom.ranks <- decostand(nullcom, method = "rank")
#     S <- specnumber(nullcom.ranks[-nrow(nullcom.ranks),])
#     nmrs <- rowSums(abs(diff(nullcom.ranks))) / S
#     mrs.null[i] <- mean(nmrs)
#     pb$tick()
#   }
#   return(mrs.null)
# }
# 
# total.null <- get.null.mrs(OTUs.total)
# active.null <- get.null.mrs(OTUs.active)
# 
# OTUs.total.ranks <- decostand(OTUs.total, method = "rank")
# MRS.total <- rowSums(abs(diff(OTUs.total.ranks))) / specnumber(OTUs.total.ranks[-nrow(OTUs.total.ranks),])
# plot(MRS.total, type = 'l')
# 
# OTUs.active.ranks <- decostand(OTUs.active, method = "rank")
# MRS.active <- rowSums(abs(diff(OTUs.active.ranks))) / specnumber(OTUs.active.ranks[-nrow(OTUs.active.ranks),])
# points(MRS.active, type = 'l', col = 'red')
# 
# MRS.total.ses <- (MRS.total - mean(total.null)) / sd(total.null)
# MRS.active.ses <- (MRS.active - mean(active.null)) / sd(active.null)
# matplot(cbind(MRS.total.ses, MRS.active.ses), type = 'l')
# 
# plot(env.vars$temp[-123], MRS.total)
# plot(env.vars$temp[-123], MRS.active)
# 
# plot(MRS.total, MRS.active)
# abline(0,1)
```

```{r}
maxgrowth <- function(sbs){
  
  # create new matrix to calculate growth rates
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  
  # for the gr's just calculated (ys) where is the max
  # for each OTU (i.e., each col) in the matrix
  max.growth.time <- apply(y.s, 2, which.max)
  
  # now, what is the gr at that max id'ed above
  max.growths <- apply(y.s, 2, max)
  
  # here, export a df of max gr and time of max gr for each OTU in the input df
  return(data.frame(
    rate = max.growths, sample = max.growth.time))
}

env.vars <- env.data[c("temp", "spc", "oxygen", "secchi", "ph", "tp", "tn", "doc")]
env.dist <- cluster::daisy(scale(env.vars), metric = "gower")
env.pca <- prcomp(na.omit(env.vars), scale. = T)

env.pcs <- as.data.frame(env.pca$x) %>% rownames_to_column(var = "sample.id") 
env.pcs$sample.id <- as.numeric(env.pcs$sample.id)
env.pcs <- env.pcs %>% right_join(env.data)

active.max.growths <- maxgrowth(OTUs.active.persistent + .1)
active.max.growths$OTU <- colnames(OTUs.active.persistent)
active.max.growths <- active.max.growths %>% mutate(sample.id = sample)

env.pcs.subset <- subset(env.pcs, sample.id %in% active.max.growths$sample.id) %>% inner_join(active.max.growths) %>%
  left_join(OTU.tax)

env.pcs.subset %>% 
  ggplot(aes(x = PC1, y = PC2)) +
  geom_jitter(aes(
    color = lubridate::month(date, label = T, abbr = T)), 
    alpha = 0.5, width = .25, height = .25, size = 2) +
  #geom_text_repel(aes(label = Order), size = 3) +
  geom_segment(data = as.data.frame(env.pca$rotation),
               aes(x = 0, y = 0, xend = 3*PC1, yend = 3*PC2),
               arrow = arrow(length = unit(.3, "cm")),
               show.legend = F, alpha = 1, color = "blue") +
  geom_text_repel(data = as.data.frame(env.pca$rotation), 
            aes(x = 3*PC1, y = 3*PC2, label = rownames(env.pca$rotation)),
            box.padding = .5, point.padding = .5, alpha = 0.9, segment.alpha = 0) +
  scale_color_viridis(discrete = T) +
  coord_fixed() +
  ggsave("figures/max_growth_envs.pdf", bg = "white", width = 8, height = 8)


getgrs <-  function(sbs){
  # create new matrix to calculate growth rates
  y.s <- matrix(ncol = ncol(sbs), nrow = nrow(sbs)-1)
  for(i in 1:(nrow(sbs)-1)){
    y.s[i,] = log(sbs[i+1,] / sbs[i,])
  }
  return(y.s)
}
persistent.grs <- getgrs(OTUs.active.persistent + .1)
colnames(persistent.grs) <- colnames(OTUs.active.persistent)
rownames(persistent.grs) <- rownames(OTUs.active.persistent)[-1]


persistent.grs.all <- as.data.frame(persistent.grs) %>% 
  rownames_to_column(var = "sample.name") %>% 
  gather(-"sample.name", key = "OTU", value = "growth") %>% 
  left_join(design) %>% 
  mutate(week = lubridate::week(date)) %>% 
  group_by(OTU, week) %>% 
  summarize(mean.growth = mean(growth),
            sd.growth = sd(growth))
maxgrs <- persistent.grs.all %>% 
  summarize(max.growth = max(mean.growth))

persistent.grs.all <- persistent.grs.all %>% 
  left_join(maxgrs) %>%
  left_join(OTU.tax)
# persistent.grs.all <- as.data.frame(persistent.grs) %>% 
#   rownames_to_column(var = "sample.name") %>% 
#   gather(-"sample.name", key = "OTU", value = "growth") %>% 
#   left_join(design) %>% 
#   left_join(OTU.tax)

persistent.grs.all <- persistent.grs.all %>% 
  mutate(ismax = ifelse(mean.growth == max.growth, T, F)) 

# this normalizes to jan 1 of each year for the avg growth rates.
persistent.grs.all$date <- lubridate::parse_date_time(paste(2014, persistent.grs.all$week, 'Wed', sep="/"),'Y/W/a') %>% as_date(.) %>% - 7
persistent.grs.all$Genus <- str_replace(persistent.grs.all$Genus, coll("_unclassified"), " sp.")

persistent.grs.all %>%   
  ggplot(aes(x = date, group = OTU)) +
  geom_line(aes(y = mean.growth), alpha = 0.1) + 
  geom_point(data = subset(persistent.grs.all, ismax == T), 
             aes(y = max.growth), alpha = 0.5, size = 3) + 
  #coord_flip() + 
  geom_ribbon(aes(group = NULL, col = NULL, ymax = 0, ymin = -5), 
              fill = "white", col = "white", alpha = 0.5) +
  theme(legend.position = "bottom", legend.text = element_text(size = 9)) +
  # geom_text_repel(data = subset(persistent.grs.all, ismax == T),
  #           aes(y = max.growth, label = Genus), 
  #           nudge_y = 0, box.padding = .5, max.iter = 10000,
  #           size = 2, color = "black", segment.alpha = .5) +
  labs(x = "", 
       y = expression(Mean~growth~rate~(wk^-1))) +
  scale_color_discrete_qualitative(palette = "Dynamic") +
  scale_x_date(date_labels = "%b", breaks = "1 month") +
  coord_polar(theta = "x") +
  theme_minimal() +
  #scale_y_continuous(limits = c(-5, 12)) +
   ggsave("figures/fig2.pdf", bg = "white", width = 8, height = 8*3/4)
  # ggsave("figures/max_growth_date.png", bg = "white", width = 12, height = 16, dpi = 600)

persistent.grs.all %>%
  filter(ismax == T) %>% 
  ggplot(aes(x = date, y = max.growth, group = OTU)) +
  geom_point() +
  labs(x = "Week of the year", 
       y = expression(Mean~growth~rate~(wk^-1))) +
  scale_color_discrete_qualitative(palette = "Dynamic") +
  scale_x_date(date_labels = "%b") +
  coord_polar(theta = "x") +
  theme_minimal()

```

